// @componlyco/berkoukes/dist/tsc/components/ConfirmationModal/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { ConfirmationModal } from "./index";
import { Button } from "../Button";
import { useState } from "react";
export default {
    title: "ConfirmationModal",
    component: ConfirmationModal,
    args: {
        title: "Title of the modal",
        description: "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aliquid ex harum laudantium porro praesentium quae reiciendis repudiandae sapiente? Amet aspernatur assumenda dignissimos esse et minus, reiciendis saepe sit tenetur voluptatibus.",
        buttonProps: {
            type: "primary",
            label: "Confirm",
        },
    },
};
export var Template = function (props) {
    var _a = useState(false), isOpened = _a[0], setOpen = _a[1];
    return (_jsxs(_Fragment, { children: [_jsx(Button, { type: "primary", label: "Modal", size: "normal", onClick: function () { return setOpen(true); } }), _jsx(ConfirmationModal, __assign({}, props, { isOpened: isOpened, onClose: function () { return setOpen(false); } }))] }));
};
export var Destructive = Template.bind({});
Destructive.args = {
    buttonProps: {
        type: "destructive",
        label: "Delete",
    },
};
export var WithImage = Template.bind({});
WithImage.args = {
    image: "https://i0.wp.com/www.flutterbeads.com/wp-content/uploads/2022/01/add-image-in-flutter-hero.png?resize=950%2C500&ssl=1",
};
