// @componlyco/berkoukes/dist/tsc/components/GitChangeList/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { GitChangeList } from "./index";
export default {
    title: "GitChangeList",
    component: GitChangeList,
    args: {
        changeList: [
            {
                title: "Update Readme.md",
                type: "git",
                role: "Engineer",
                branch: "test-project",
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
            {
                title: "Update Readme.md",
                type: "git",
                role: "Engineer",
                branch: "test-project",
                disabled: true,
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
            {
                title: "Change ratio from 12px to 16px",
                type: "figma",
                role: "Design",
                branch: "main@b2988a6",
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
            {
                title: "Update Readme.md",
                type: "git",
                role: "Engineer",
                branch: "test-project",
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
            {
                title: "Update Readme.md",
                type: "git",
                role: "Engineer",
                branch: "test-project",
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
            {
                title: "Update Readme.md",
                type: "git",
                role: "Engineer",
                branch: "test-project",
                user_profile: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            },
        ],
        sx: {
            minWidth: 700,
        },
    },
};
var Template = function (args) { return (_jsx(GitChangeList, __assign({}, args))); };
export var Default = Template.bind({});
export var White = Template.bind({});
White.args = {
    background: "#ffffff",
    sx: {
        borderRadius: 30,
    },
};
