// @componlyco/berkoukes/dist/tsc/components/Select/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ActionIcon, Flex, Group, Select as MantineSelect, Stack, Text } from "@mantine/core";
import { ChevronDown, HelpCircle } from "lucide-react";
import { BoxShadows, Colors, FontSizes, Layout } from "../../design-tokens";
import { Tooltip } from "../Tooltip";
export function Select(_a) {
    var label = _a.label, data = _a.data, icon = _a.icon, onHelpIconClick = _a.onHelpIconClick, optionalText = _a.optionalText, error = _a.error, helperText = _a.helperText, placeholder = _a.placeholder, disabled = _a.disabled, defaultValue = _a.defaultValue, sx = _a.sx, onChange = _a.onChange, value = _a.value, onBlur = _a.onBlur, _b = _a.name, name = _b === void 0 ? undefined : _b, _c = _a.showHelpIcon, showHelpIcon = _c === void 0 ? true : _c, _d = _a.helpTooltipProps, helpTooltipProps = _d === void 0 ? {
        children: null,
        label: null,
        position: "top",
        title: "Header",
        description: "In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content.",
    } : _d, selector = _a.selector, rest = __rest(_a, ["label", "data", "icon", "onHelpIconClick", "optionalText", "error", "helperText", "placeholder", "disabled", "defaultValue", "sx", "onChange", "value", "onBlur", "name", "showHelpIcon", "helpTooltipProps", "selector"]);
    return (_jsxs(Stack, __assign({ spacing: Layout.sm, sx: sx }, { children: [_jsxs(Flex, __assign({ direction: "row", align: "center", justify: "space-between" }, { children: [_jsxs(Group, __assign({ spacing: Layout.xs, align: "center" }, { children: [_jsx(Text, __assign({ sx: __assign({}, FontSizes.Subtitle2) }, { children: label })), showHelpIcon && (_jsx(Tooltip, __assign({}, helpTooltipProps, { children: _jsx(ActionIcon, __assign({ variant: "subtle", size: Layout.TBR, onClick: onHelpIconClick, color: "gray", radius: "xl" }, { children: _jsx(HelpCircle, { size: "19px", style: {
                                            fill: Colors.Neutral[500],
                                            stroke: "white",
                                        } }) })) })))] })), typeof optionalText === "string" ? (_jsx(Text, __assign({ color: Colors.DarkGrey[600], sx: __assign({}, FontSizes.Helper) }, { children: optionalText }))) : undefined] })), _jsx(MantineSelect, __assign({ icon: icon, placeholder: placeholder, disabled: disabled, onChange: onChange, onBlur: onBlur, error: error, name: name, data: data, styles: {
                    label: __assign({}, FontSizes.Subtitle2),
                    error: __assign(__assign({}, FontSizes.Helper), { color: Colors.Badge.error.Label }),
                    input: __assign(__assign({}, FontSizes.Body2), { "&[aria-invalid='false']:focus": {
                            border: "1.5px solid ".concat(Colors.Default.Background),
                            boxShadow: BoxShadows.Input,
                        }, "&[aria-invalid='true']:focus": {
                            border: "1.5px solid ".concat(Colors.Button.destructive.Hover.Text),
                        } }),
                    item: __assign(__assign({}, FontSizes.Body2), { borderRadius: "30px" }),
                    dropdown: {
                        borderRadius: "30px",
                    },
                    /**
                     * Will make right section icon to not be a target of any pointer events,
                     * thus whole 'MultiSelect' will get precedance over 'pointer' events.
                     * Ref: https://mantine.dev/core/multi-select/#right-section
                     */
                    rightSection: { pointerEvents: "none" },
                }, radius: "lg", defaultValue: defaultValue, value: value, rightSection: _jsx(ChevronDown, {}) }, selector, rest)), typeof error !== "string" ? (_jsx(Text, __assign({ color: Colors.DarkGrey[600], sx: __assign({}, FontSizes.Helper) }, { children: helperText }))) : undefined] })));
}
