// @componlyco/berkoukes/dist/tsc/tokens/Colors/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { createStyles, Stack, Text } from "@mantine/core";
import { ColorsToken } from ".";
import { Colors } from "../../design-tokens";
import logo from '../../assets/logo.svg';
import { withDesign } from "storybook-addon-designs";
export default {
    title: "tokens/Colors",
    component: ColorsToken,
    argTypes: {
        position: {
            control: "select",
        },
    },
    args: {
        position: "top",
        title: "Header",
        description: "In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content.",
    },
    decorators: [withDesign],
    parameters: {
        design: {
            type: "figma",
            // url: "https://www.figma.com/file/LKQ4FJ4bTnCSjedbRpk931/Sample-File",
            url: "https://www.figma.com/file/YfE88cxwKQfAXloAsAV3Bv/Berkoukes-DS?type=design&node-id=67-100602&mode=dev",
        },
    }
};
var palette = {
    main: [
        [
            '#E6F2FF',
            "#96C9FF",
            "#6BB2FF",
            "#2B91FF",
            "#007BFF",
            "#0056B3",
            "#004B9C",
        ],
        [
            "#FFFFFF",
            "#F5F6F8",
            "#F5F6F8",
            "#E5E5E6",
            "#E0E1E2",
            "#C3C5C8",
            "#A9ABB0",
        ],
        [
            "#9E9E9E",
            "#757575",
            "#616161",
            "#404040",
            "#212121",
            "#000000",
        ],
        [
            "#EDEEF0",
            "#B4B9C1",
            "#949CA7",
            "#667181",
            "#475467",
            "#323B48",
            "#2B333F",
        ],
        [
            "#FAFBFC",
            "#ECEEF1",
            "#E4E7EB",
            "#D8DCE3",
            "#D0D5DD",
            "#92959B",
            "#7F8287",
        ],
        [
            "#E7F5EC",
            "#9ED5B1",
            "#76C391",
            "#3BAA61",
            "#139841",
            "#0D6A2E",
            "#0C5D28",
        ],
        [
            "#FDF5EA",
            "#F7D4A8",
            "#F3C383",
            "#EEA94E",
            "#EB972A",
            "#A56A1D",
            "#8F5C1A"
        ],
        [
            "#FCE9E9",
            "#F2A3A6",
            "#ED7D82",
            "#E5444C",
            "#E01E27",
            "#9D151B",
            "#891218"
        ]
    ],
    chart: [
        [
            "#FAFCFB",
            "#EAF1ED",
            "#E1EBE5",
            "#D5E3DA",
            "#CCDDD3",
            "#8F9B94",
            "#7C8781"
        ],
        [
            "#E9EBEC",
            "#A3ADAF",
            "#7D8A8E",
            "#45585E",
            "#1F363D",
            "#16262B",
            "#132125",
        ],
        [
            "#F1EBFF",
            "#C6ADFF",
            "#AE8CFE",
            "#8C5AFE",
            "#7438FE",
            "#5127B2",
            "#47229B",
        ],
        [
            "#FAF6FF",
            "#1F363D",
            "#7438FE",
            "#CEA9FE",
            "#FED4E7",
            "#EA9E8D",
            "#FC4B57",
        ]
    ],
    logo: [
        [
            "#C4C4C4",
            "#FF7262"
        ]
    ]
};
var useStyles = createStyles(function () { return ({
    root: {
        width: "calc(100vw - 32px)",
        gap: "54px",
        flexDirection: "column"
    },
    header: {
        background: "#E6F2FF",
        borderRadius: "17px",
        display: "flex",
        padding: "40px",
        flexDirection: "column",
        alignItems: "flex-start",
        gap: "32px",
        alignSelf: "stretch"
    },
    groupRoot: {
        flexDirection: "column",
        alignSelf: "stretch",
        gap: "0",
        borderRadius: "10px",
        border: "1px solid #D0D5DD",
        background: "#FFFFFF"
    },
    groupHead: { padding: "24px" },
    groupBody: { background: "#F5F6F8" },
    subgroupBody: {
        padding: "48px 24px",
        flexDirection: "row",
        gap: "24px",
        flexWrap: "wrap"
    },
    color: {
        width: "200px",
        height: "100px",
        borderRadius: "30px",
        border: "0.5px solid #000000"
    }
}); });
var Template = function (args) {
    var s = useStyles().classes;
    return (_jsxs(Stack, __assign({ className: s.root }, { children: [_jsxs(Stack, __assign({ className: s.header, spacing: "32px" }, { children: [_jsxs(Stack, __assign({ justify: "space-between", align: "center", sx: { flexDirection: "row", width: "100%" } }, { children: [_jsxs(Stack, __assign({ spacing: "12px", align: "center", sx: { flexDirection: "row" } }, { children: [_jsx(HexIcon, {}), _jsx(Text, __assign({ fz: 18, color: Colors.Primary.blue }, { children: "Design system" }))] })), _jsx("img", { src: logo, alt: "logo", height: "29px" })] })), _jsx(Text, __assign({ fz: 54 }, { children: "Colors" }))] })), _jsxs(Stack, __assign({ className: s.groupRoot }, { children: [_jsx(Stack, __assign({ className: s.groupHead }, { children: _jsx(Text, __assign({ fz: 24 }, { children: "Main" })) })), _jsx(Stack, __assign({ className: s.groupBody }, { children: palette.main.map(function (g, i) {
                            return _jsx(Stack, __assign({ className: s.subgroupBody, spacing: "24px" }, { children: g === null || g === void 0 ? void 0 : g.map(function (c, i) { return _jsx("div", { className: s.color, style: { backgroundColor: c } }, i); }) }), i);
                        }) }))] })), _jsxs(Stack, __assign({ className: s.groupRoot }, { children: [_jsx(Stack, __assign({ className: s.groupHead }, { children: _jsx(Text, __assign({ fz: 24 }, { children: "Charts Color" })) })), _jsx(Stack, __assign({ className: s.groupBody }, { children: palette.chart.map(function (g, i) {
                            return _jsx(Stack, __assign({ className: s.subgroupBody, spacing: "24px" }, { children: g === null || g === void 0 ? void 0 : g.map(function (c, i) { return _jsx("div", { className: s.color, style: { backgroundColor: c } }, i); }) }), i);
                        }) }))] })), _jsxs(Stack, __assign({ className: s.groupRoot }, { children: [_jsx(Stack, __assign({ className: s.groupHead }, { children: _jsx(Text, __assign({ fz: 24 }, { children: "Logo" })) })), _jsx(Stack, __assign({ className: s.groupBody }, { children: palette.logo.map(function (g, i) {
                            return _jsx(Stack, __assign({ className: s.subgroupBody, spacing: "24px" }, { children: g === null || g === void 0 ? void 0 : g.map(function (c, i) { return _jsx("div", { className: s.color, style: { backgroundColor: c } }, i); }) }), i);
                        }) }))] })), _jsx(ColorsToken, __assign({}, args))] })));
};
export var Default = Template.bind({});
var HexIcon = function () {
    return _jsxs("svg", __assign({ width: "16", height: "17", viewBox: "0 0 16 17", fill: "none", xmlns: "http://www.w3.org/2000/svg" }, { children: [_jsx("path", { d: "M7.92578 0.507812L11.4456 4.02763L7.92578 7.54745L4.40596 4.02763L7.92578 0.507812Z", fill: "#007BFF" }), _jsx("path", { d: "M12.3256 4.90759L15.8454 8.42741L12.3256 11.9472L8.80574 8.42741L12.3256 4.90759Z", fill: "#007BFF" }), _jsx("path", { d: "M3.52601 4.90759L7.04583 8.42741L3.52601 11.9472L0.0061849 8.42741L3.52601 4.90759Z", fill: "#007BFF" }), _jsx("path", { d: "M7.92578 9.30736L11.4456 12.8272L7.92578 16.347L4.40596 12.8272L7.92578 9.30736Z", fill: "#007BFF" })] }));
};
