// @componlyco/berkoukes/dist/tsc/components/VerificationInput/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { VerificationInput } from "./index";
export default {
    title: "VerificationInput",
    component: VerificationInput,
    args: {
        label: "Label",
        optionalText: "Optional",
        helperText: "Helper Text",
    },
};
var Template = function (args) { return (_jsx(VerificationInput, __assign({}, args))); };
export var Default = Template.bind({});
export var Error = Template.bind({});
Error.args = {
    error: true,
};
export var WithoutOptional = Template.bind({});
WithoutOptional.args = {
    optionalText: undefined,
};
export var Large = Template.bind({});
Large.args = {
    length: 8,
    helperText: "Helper text",
};
