// @componlyco/berkoukes/dist/tsc/components/VerificationInput/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { PinInput, Text, Stack, ActionIcon, Flex, Group } from "@mantine/core";
import { HelpCircle } from "lucide-react";
import { BoxShadows, Colors, FontSizes, Layout } from "../../design-tokens";
import { Tooltip } from "../Tooltip";
export function VerificationInput(_a) {
    var value = _a.value, length = _a.length, onChange = _a.onChange, onComplete = _a.onComplete, error = _a.error, helperText = _a.helperText, label = _a.label, sx = _a.sx, onHelpIconClick = _a.onHelpIconClick, optionalText = _a.optionalText, _b = _a.showHelpIcon, showHelpIcon = _b === void 0 ? true : _b, _c = _a.helpTooltipProps, helpTooltipProps = _c === void 0 ? {
        children: null,
        label: null,
        position: "top",
        title: "Header",
        description: "In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content.",
    } : _c, selector = _a.selector;
    return (_jsxs(Stack, __assign({ sx: sx, spacing: Layout.sm }, { children: [_jsxs(Flex, __assign({ direction: "row", align: "center", justify: "space-between" }, { children: [_jsxs(Group, __assign({ spacing: Layout.xs, align: "center" }, { children: [_jsx(Text, __assign({ sx: __assign({}, FontSizes.Subtitle2) }, { children: label })), showHelpIcon ? (_jsx(Tooltip, __assign({}, helpTooltipProps, { children: _jsx(ActionIcon, __assign({ size: Layout.TBR, variant: "subtle", onClick: onHelpIconClick, color: "gray", radius: "xl" }, { children: _jsx(HelpCircle, { size: "19px", style: {
                                            fill: Colors.Neutral[500],
                                            stroke: "white",
                                        } }) })) }))) : null] })), typeof optionalText === "string" ? (_jsx(Text, __assign({ color: Colors.DarkGrey[600], sx: __assign({}, FontSizes.Helper) }, { children: optionalText }))) : undefined] })), _jsx(PinInput, __assign({ value: value, length: length !== null && length !== void 0 ? length : 4, onChange: onChange, onComplete: onComplete, error: error, placeholder: "-", radius: "xl", size: "xl", styles: {
                    root: {
                        gap: Layout.md
                    },
                    input: __assign(__assign({}, FontSizes.Body1), { "&[aria-invalid='false']:focus": {
                            border: "1.5px solid ".concat(Colors.Default.Background),
                            boxShadow: BoxShadows.Input,
                        }, "&[aria-invalid='true']:focus": {
                            border: "1.5px solid ".concat(Colors.Button.destructive.Hover.Text),
                        }, boxShadow: BoxShadows.VerificationInput }),
                } }, selector)), typeof helperText === "string" ? (_jsx(Text, __assign({ sx: __assign({}, FontSizes.Helper), color: error === true ? "red" : Colors.DarkGrey[600] }, { children: helperText }))) : null] })));
}
