// @componlyco/core/dist/tsc/parse/angular/compiler/src/template/pipeline/ir/src/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BindingKind = exports.SanitizerFn = exports.CompatibilityMode = exports.SemanticVariableKind = exports.ExpressionKind = exports.OpKind = void 0;
var OpKind;
(function (OpKind) {
    OpKind[OpKind["ListEnd"] = 0] = "ListEnd";
    OpKind[OpKind["Statement"] = 1] = "Statement";
    OpKind[OpKind["Variable"] = 2] = "Variable";
    OpKind[OpKind["ElementStart"] = 3] = "ElementStart";
    OpKind[OpKind["Element"] = 4] = "Element";
    OpKind[OpKind["Template"] = 5] = "Template";
    OpKind[OpKind["ElementEnd"] = 6] = "ElementEnd";
    OpKind[OpKind["ContainerStart"] = 7] = "ContainerStart";
    OpKind[OpKind["Container"] = 8] = "Container";
    OpKind[OpKind["ContainerEnd"] = 9] = "ContainerEnd";
    OpKind[OpKind["DisableBindings"] = 10] = "DisableBindings";
    OpKind[OpKind["EnableBindings"] = 11] = "EnableBindings";
    OpKind[OpKind["Text"] = 12] = "Text";
    OpKind[OpKind["Listener"] = 13] = "Listener";
    OpKind[OpKind["InterpolateText"] = 14] = "InterpolateText";
    OpKind[OpKind["Binding"] = 15] = "Binding";
    OpKind[OpKind["Property"] = 16] = "Property";
    OpKind[OpKind["StyleProp"] = 17] = "StyleProp";
    OpKind[OpKind["ClassProp"] = 18] = "ClassProp";
    OpKind[OpKind["StyleMap"] = 19] = "StyleMap";
    OpKind[OpKind["ClassMap"] = 20] = "ClassMap";
    OpKind[OpKind["Advance"] = 21] = "Advance";
    OpKind[OpKind["Pipe"] = 22] = "Pipe";
    OpKind[OpKind["Attribute"] = 23] = "Attribute";
    OpKind[OpKind["ExtractedAttribute"] = 24] = "ExtractedAttribute";
    OpKind[OpKind["HostProperty"] = 25] = "HostProperty";
    OpKind[OpKind["Namespace"] = 26] = "Namespace";
})(OpKind = exports.OpKind || (exports.OpKind = {}));
var ExpressionKind;
(function (ExpressionKind) {
    ExpressionKind[ExpressionKind["LexicalRead"] = 0] = "LexicalRead";
    ExpressionKind[ExpressionKind["Context"] = 1] = "Context";
    ExpressionKind[ExpressionKind["ReadVariable"] = 2] = "ReadVariable";
    ExpressionKind[ExpressionKind["NextContext"] = 3] = "NextContext";
    ExpressionKind[ExpressionKind["Reference"] = 4] = "Reference";
    ExpressionKind[ExpressionKind["GetCurrentView"] = 5] = "GetCurrentView";
    ExpressionKind[ExpressionKind["RestoreView"] = 6] = "RestoreView";
    ExpressionKind[ExpressionKind["ResetView"] = 7] = "ResetView";
    ExpressionKind[ExpressionKind["PureFunctionExpr"] = 8] = "PureFunctionExpr";
    ExpressionKind[ExpressionKind["PureFunctionParameterExpr"] = 9] = "PureFunctionParameterExpr";
    ExpressionKind[ExpressionKind["PipeBinding"] = 10] = "PipeBinding";
    ExpressionKind[ExpressionKind["PipeBindingVariadic"] = 11] = "PipeBindingVariadic";
    ExpressionKind[ExpressionKind["SafePropertyRead"] = 12] = "SafePropertyRead";
    ExpressionKind[ExpressionKind["SafeKeyedRead"] = 13] = "SafeKeyedRead";
    ExpressionKind[ExpressionKind["SafeInvokeFunction"] = 14] = "SafeInvokeFunction";
    ExpressionKind[ExpressionKind["SafeTernaryExpr"] = 15] = "SafeTernaryExpr";
    ExpressionKind[ExpressionKind["EmptyExpr"] = 16] = "EmptyExpr";
    ExpressionKind[ExpressionKind["AssignTemporaryExpr"] = 17] = "AssignTemporaryExpr";
    ExpressionKind[ExpressionKind["ReadTemporaryExpr"] = 18] = "ReadTemporaryExpr";
    ExpressionKind[ExpressionKind["SanitizerExpr"] = 19] = "SanitizerExpr";
})(ExpressionKind = exports.ExpressionKind || (exports.ExpressionKind = {}));
var SemanticVariableKind;
(function (SemanticVariableKind) {
    SemanticVariableKind[SemanticVariableKind["Context"] = 0] = "Context";
    SemanticVariableKind[SemanticVariableKind["Identifier"] = 1] = "Identifier";
    SemanticVariableKind[SemanticVariableKind["SavedView"] = 2] = "SavedView";
})(SemanticVariableKind = exports.SemanticVariableKind || (exports.SemanticVariableKind = {}));
var CompatibilityMode;
(function (CompatibilityMode) {
    CompatibilityMode[CompatibilityMode["Normal"] = 0] = "Normal";
    CompatibilityMode[CompatibilityMode["TemplateDefinitionBuilder"] = 1] = "TemplateDefinitionBuilder";
})(CompatibilityMode = exports.CompatibilityMode || (exports.CompatibilityMode = {}));
var SanitizerFn;
(function (SanitizerFn) {
    SanitizerFn[SanitizerFn["Html"] = 0] = "Html";
    SanitizerFn[SanitizerFn["Script"] = 1] = "Script";
    SanitizerFn[SanitizerFn["Style"] = 2] = "Style";
    SanitizerFn[SanitizerFn["Url"] = 3] = "Url";
    SanitizerFn[SanitizerFn["ResourceUrl"] = 4] = "ResourceUrl";
    SanitizerFn[SanitizerFn["IframeAttribute"] = 5] = "IframeAttribute";
})(SanitizerFn = exports.SanitizerFn || (exports.SanitizerFn = {}));
var BindingKind;
(function (BindingKind) {
    BindingKind[BindingKind["Attribute"] = 0] = "Attribute";
    BindingKind[BindingKind["ClassName"] = 1] = "ClassName";
    BindingKind[BindingKind["StyleProperty"] = 2] = "StyleProperty";
    BindingKind[BindingKind["Property"] = 3] = "Property";
    BindingKind[BindingKind["Template"] = 4] = "Template";
    BindingKind[BindingKind["I18n"] = 5] = "I18n";
    BindingKind[BindingKind["Animation"] = 6] = "Animation";
})(BindingKind = exports.BindingKind || (exports.BindingKind = {}));
