// @componlyco/core/dist/tsc/parse/angular/compiler/src/template/pipeline/ir/src/traits.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasUsesSlotIndexTrait = exports.hasUsesVarOffsetTrait = exports.hasConsumesVarsTrait = exports.hasDependsOnSlotContextTrait = exports.hasConsumesSlotTrait = exports.TRAIT_USES_VAR_OFFSET = exports.TRAIT_CONSUMES_VARS = exports.TRAIT_DEPENDS_ON_SLOT_CONTEXT = exports.TRAIT_USES_SLOT_INDEX = exports.TRAIT_CONSUMES_SLOT = exports.UsesVarOffset = exports.ConsumesVarsTrait = exports.UsesSlotIndex = exports.DependsOnSlotContext = exports.ConsumesSlot = void 0;
exports.ConsumesSlot = Symbol('ConsumesSlot');
exports.DependsOnSlotContext = Symbol('DependsOnSlotContext');
exports.UsesSlotIndex = Symbol('UsesSlotIndex');
exports.ConsumesVarsTrait = Symbol('ConsumesVars');
exports.UsesVarOffset = Symbol('UsesVarOffset');
exports.TRAIT_CONSUMES_SLOT = (_a = {},
    _a[exports.ConsumesSlot] = true,
    _a.slot = null,
    _a.numSlotsUsed = 1,
    _a);
exports.TRAIT_USES_SLOT_INDEX = (_b = {},
    _b[exports.UsesSlotIndex] = true,
    _b.slot = null,
    _b);
exports.TRAIT_DEPENDS_ON_SLOT_CONTEXT = (_c = {},
    _c[exports.DependsOnSlotContext] = true,
    _c);
exports.TRAIT_CONSUMES_VARS = (_d = {},
    _d[exports.ConsumesVarsTrait] = true,
    _d);
exports.TRAIT_USES_VAR_OFFSET = (_e = {},
    _e[exports.UsesVarOffset] = true,
    _e.varOffset = null,
    _e);
function hasConsumesSlotTrait(op) {
    return op[exports.ConsumesSlot] === true;
}
exports.hasConsumesSlotTrait = hasConsumesSlotTrait;
function hasDependsOnSlotContextTrait(op) {
    return op[exports.DependsOnSlotContext] === true;
}
exports.hasDependsOnSlotContextTrait = hasDependsOnSlotContextTrait;
function hasConsumesVarsTrait(value) {
    return value[exports.ConsumesVarsTrait] === true;
}
exports.hasConsumesVarsTrait = hasConsumesVarsTrait;
function hasUsesVarOffsetTrait(expr) {
    return expr[exports.UsesVarOffset] === true;
}
exports.hasUsesVarOffsetTrait = hasUsesVarOffsetTrait;
function hasUsesSlotIndexTrait(value) {
    return value[exports.UsesSlotIndex] === true;
}
exports.hasUsesSlotIndexTrait = hasUsesSlotIndexTrait;
