// @componlyco/core/dist/tsc/parse/angular/compiler/src/template/pipeline/src/phases/const_collection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.phaseConstCollection = void 0;
var tags_1 = require("../../../../ml_parser/tags");
var o = __importStar(require("../../../../output/output_ast"));
var ir = __importStar(require("../../ir"));
function phaseConstCollection(cpl) {
    var e_1, _a, e_2, _b, e_3, _c, e_4, _d;
    var elementAttributes = new Map();
    try {
        for (var _e = __values(cpl.views), _f = _e.next(); !_f.done; _f = _e.next()) {
            var _g = __read(_f.value, 2), _1 = _g[0], view = _g[1];
            try {
                for (var _h = (e_2 = void 0, __values(view.create)), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var op = _j.value;
                    if (op.kind === ir.OpKind.ExtractedAttribute) {
                        var attributes = elementAttributes.get(op.target) || new ElementAttributes();
                        elementAttributes.set(op.target, attributes);
                        attributes.add(op.bindingKind, op.name, op.expression);
                        ir.OpList.remove(op);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
        }
        finally { if (e_1) throw e_1.error; }
    }
    try {
        for (var _k = __values(cpl.views), _l = _k.next(); !_l.done; _l = _k.next()) {
            var _m = __read(_l.value, 2), _2 = _m[0], view = _m[1];
            try {
                for (var _o = (e_4 = void 0, __values(view.create)), _p = _o.next(); !_p.done; _p = _o.next()) {
                    var op = _p.value;
                    if (op.kind === ir.OpKind.Element || op.kind === ir.OpKind.ElementStart ||
                        op.kind === ir.OpKind.Template) {
                        var attributes = elementAttributes.get(op.xref);
                        if (attributes !== undefined) {
                            var attrArray = serializeAttributes(attributes);
                            if (attrArray.entries.length > 0) {
                                op.attributes = cpl.addConst(attrArray);
                            }
                        }
                    }
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_p && !_p.done && (_d = _o.return)) _d.call(_o);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
        }
        finally { if (e_3) throw e_3.error; }
    }
}
exports.phaseConstCollection = phaseConstCollection;
var FLYWEIGHT_ARRAY = Object.freeze([]);
var ElementAttributes = (function () {
    function ElementAttributes() {
        this.known = new Set();
        this.byKind = new Map;
        this.projectAs = null;
    }
    Object.defineProperty(ElementAttributes.prototype, "attributes", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.Attribute)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ElementAttributes.prototype, "classes", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.ClassName)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ElementAttributes.prototype, "styles", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.StyleProperty)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ElementAttributes.prototype, "bindings", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.Property)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ElementAttributes.prototype, "template", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.Template)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ElementAttributes.prototype, "i18n", {
        get: function () {
            var _a;
            return (_a = this.byKind.get(ir.BindingKind.I18n)) !== null && _a !== void 0 ? _a : FLYWEIGHT_ARRAY;
        },
        enumerable: false,
        configurable: true
    });
    ElementAttributes.prototype.add = function (kind, name, value) {
        if (this.known.has(name)) {
            return;
        }
        this.known.add(name);
        var array = this.arrayFor(kind);
        array.push.apply(array, __spreadArray([], __read(getAttributeNameLiterals(name)), false));
        if (kind === ir.BindingKind.Attribute || kind === ir.BindingKind.StyleProperty) {
            if (value === null) {
                throw Error('Attribute & style element attributes must have a value');
            }
            array.push(value);
        }
    };
    ElementAttributes.prototype.arrayFor = function (kind) {
        if (!this.byKind.has(kind)) {
            this.byKind.set(kind, []);
        }
        return this.byKind.get(kind);
    };
    return ElementAttributes;
}());
function getAttributeNameLiterals(name) {
    var _a = __read((0, tags_1.splitNsName)(name), 2), attributeNamespace = _a[0], attributeName = _a[1];
    var nameLiteral = o.literal(attributeName);
    if (attributeNamespace) {
        return [
            o.literal(0), o.literal(attributeNamespace), nameLiteral
        ];
    }
    return [nameLiteral];
}
function serializeAttributes(_a) {
    var attributes = _a.attributes, bindings = _a.bindings, classes = _a.classes, i18n = _a.i18n, projectAs = _a.projectAs, styles = _a.styles, template = _a.template;
    var attrArray = __spreadArray([], __read(attributes), false);
    if (projectAs !== null) {
        attrArray.push(o.literal(5), o.literal(projectAs));
    }
    if (classes.length > 0) {
        attrArray.push.apply(attrArray, __spreadArray([o.literal(1)], __read(classes), false));
    }
    if (styles.length > 0) {
        attrArray.push.apply(attrArray, __spreadArray([o.literal(2)], __read(styles), false));
    }
    if (bindings.length > 0) {
        attrArray.push.apply(attrArray, __spreadArray([o.literal(3)], __read(bindings), false));
    }
    if (template.length > 0) {
        attrArray.push.apply(attrArray, __spreadArray([o.literal(4)], __read(template), false));
    }
    if (i18n.length > 0) {
        attrArray.push.apply(attrArray, __spreadArray([o.literal(6)], __read(i18n), false));
    }
    return o.literalArr(attrArray);
}
