// @componlyco/berkoukes/dist/tsc/components/Avatar/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Avatar } from "./index";
export default {
    title: "Avatar",
    component: Avatar,
};
var Template = function (args) { return _jsx(Avatar, __assign({}, args)); };
export var Circle = Template.bind({});
Circle.args = {
    type: "image",
    size: "xxl",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
export var Square = Template.bind({});
Square.args = {
    shape: "square",
    type: "image",
    size: "xxl",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
export var TextCircle = Template.bind({});
TextCircle.args = {
    type: "text",
    letter: "M",
    size: "xxl",
};
export var AvatarCircle = Template.bind({});
AvatarCircle.args = {
    type: "avatar",
    size: "xxl",
};
export var ImageXL = Template.bind({});
ImageXL.args = {
    type: "image",
    size: "xxl",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
export var ImageLarge = Template.bind({});
ImageLarge.args = {
    type: "image",
    size: "xl",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
export var ImageMedium = Template.bind({});
ImageMedium.args = {
    type: "image",
    size: "l",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
export var ImageSmall = Template.bind({});
ImageSmall.args = {
    type: "image",
    size: "m",
    src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
};
