// @componlyco/berkoukes/dist/tsc/components/SideBarNav/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { BarChart, FolderArchiveIcon, UserIcon, Hexagon, Home, SettingsIcon } from "lucide-react";
import settingsIcon from "../../assets/settings-icon.svg";
import { ButtonType } from "../Button";
import { UserDropdownTypes } from "../UserProfile";
import { SideBarNav } from "./index";
export default {
    title: "SideBarNav",
    component: SideBarNav,
    args: {
        title: 'Heading',
        collapse: true,
        menuItems: [
            {
                type: ButtonType.NoBorders,
                label: "Home",
                icon: _jsx(Home, {}),
            },
            {
                type: ButtonType.NoBorders,
                label: "Dashboards",
                icon: _jsx(Hexagon, {}),
                current: true,
            },
            {
                type: ButtonType.NoBorders,
                label: "Users",
                icon: _jsx(UserIcon, {}),
            },
            {
                type: ButtonType.NoBorders,
                label: "Projects",
                icon: _jsx(FolderArchiveIcon, {}),
            },
            {
                type: ButtonType.NoBorders,
                label: "Projects",
                icon: _jsx(BarChart, {}),
            },
            {
                type: ButtonType.Borders,
                icon: _jsx("img", { src: settingsIcon, alt: "settingsIcon" }),
                sx: {
                    marginTop: "auto",
                    justifyContent: "center",
                },
            },
        ],
        userProfile: {
            src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            name: "Team name",
            handle: "@teamhandle",
            type: UserDropdownTypes.Contained,
            avatarType: "image",
            avatarSize: 40,
            width: "calc( 100% - 100px )",
            items: [
                { label: "Resources", onClick: function () { return alert("Resources"); } },
                { label: "Billing" },
                { label: "Settings" },
                { label: "Logout", icon: _jsx(SettingsIcon, {}) },
            ]
        },
    },
};
var Template = function (args) { return _jsx(SideBarNav, __assign({}, args)); };
export var Collapse = Template.bind({});
export var Expand = Template.bind({});
Expand.args = {
    collapse: false,
};
