// @componlyco/core/dist/tsc/parse/angular/compiler/src/template/pipeline/src/compilation.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViewCompilationUnit = exports.ComponentCompilationJob = exports.HostBindingCompilationJob = exports.CompilationUnit = void 0;
var ir = __importStar(require("../ir"));
var CompilationUnit = (function () {
    function CompilationUnit(xref) {
        this.xref = xref;
        this.create = new ir.OpList();
        this.update = new ir.OpList();
        this.fnName = null;
        this.vars = null;
    }
    CompilationUnit.prototype.ops = function () {
        var _a, _b, op, _c, _d, listenerOp, e_1_1, e_2_1, _e, _f, op, e_3_1;
        var e_2, _g, e_1, _h, e_3, _j;
        return __generator(this, function (_k) {
            switch (_k.label) {
                case 0:
                    _k.trys.push([0, 12, 13, 14]);
                    _a = __values(this.create), _b = _a.next();
                    _k.label = 1;
                case 1:
                    if (!!_b.done) return [3, 11];
                    op = _b.value;
                    return [4, op];
                case 2:
                    _k.sent();
                    if (!(op.kind === ir.OpKind.Listener)) return [3, 10];
                    _k.label = 3;
                case 3:
                    _k.trys.push([3, 8, 9, 10]);
                    _c = (e_1 = void 0, __values(op.handlerOps)), _d = _c.next();
                    _k.label = 4;
                case 4:
                    if (!!_d.done) return [3, 7];
                    listenerOp = _d.value;
                    return [4, listenerOp];
                case 5:
                    _k.sent();
                    _k.label = 6;
                case 6:
                    _d = _c.next();
                    return [3, 4];
                case 7: return [3, 10];
                case 8:
                    e_1_1 = _k.sent();
                    e_1 = { error: e_1_1 };
                    return [3, 10];
                case 9:
                    try {
                        if (_d && !_d.done && (_h = _c.return)) _h.call(_c);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7];
                case 10:
                    _b = _a.next();
                    return [3, 1];
                case 11: return [3, 14];
                case 12:
                    e_2_1 = _k.sent();
                    e_2 = { error: e_2_1 };
                    return [3, 14];
                case 13:
                    try {
                        if (_b && !_b.done && (_g = _a.return)) _g.call(_a);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7];
                case 14:
                    _k.trys.push([14, 19, 20, 21]);
                    _e = __values(this.update), _f = _e.next();
                    _k.label = 15;
                case 15:
                    if (!!_f.done) return [3, 18];
                    op = _f.value;
                    return [4, op];
                case 16:
                    _k.sent();
                    _k.label = 17;
                case 17:
                    _f = _e.next();
                    return [3, 15];
                case 18: return [3, 21];
                case 19:
                    e_3_1 = _k.sent();
                    e_3 = { error: e_3_1 };
                    return [3, 21];
                case 20:
                    try {
                        if (_f && !_f.done && (_j = _e.return)) _j.call(_e);
                    }
                    finally { if (e_3) throw e_3.error; }
                    return [7];
                case 21: return [2];
            }
        });
    };
    return CompilationUnit;
}());
exports.CompilationUnit = CompilationUnit;
var HostBindingCompilationJob = (function (_super) {
    __extends(HostBindingCompilationJob, _super);
    function HostBindingCompilationJob(componentName, pool, compatibility) {
        var _this = _super.call(this, 0) || this;
        _this.componentName = componentName;
        _this.pool = pool;
        _this.compatibility = compatibility;
        _this.fnSuffix = 'HostBindings';
        _this.units = [_this];
        _this.nextXrefId = 1;
        return _this;
    }
    Object.defineProperty(HostBindingCompilationJob.prototype, "job", {
        get: function () {
            return this;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(HostBindingCompilationJob.prototype, "root", {
        get: function () {
            return this;
        },
        enumerable: false,
        configurable: true
    });
    HostBindingCompilationJob.prototype.allocateXrefId = function () {
        return this.nextXrefId++;
    };
    return HostBindingCompilationJob;
}(CompilationUnit));
exports.HostBindingCompilationJob = HostBindingCompilationJob;
var ComponentCompilationJob = (function () {
    function ComponentCompilationJob(componentName, pool, compatibility) {
        this.componentName = componentName;
        this.pool = pool;
        this.compatibility = compatibility;
        this.fnSuffix = 'Template';
        this.nextXrefId = 0;
        this.views = new Map();
        this.consts = [];
        var root = new ViewCompilationUnit(this, this.allocateXrefId(), null);
        this.views.set(root.xref, root);
        this.root = root;
    }
    Object.defineProperty(ComponentCompilationJob.prototype, "units", {
        get: function () {
            return this.views.values();
        },
        enumerable: false,
        configurable: true
    });
    ComponentCompilationJob.prototype.allocateView = function (parent) {
        var view = new ViewCompilationUnit(this, this.allocateXrefId(), parent);
        this.views.set(view.xref, view);
        return view;
    };
    ComponentCompilationJob.prototype.allocateXrefId = function () {
        return this.nextXrefId++;
    };
    ComponentCompilationJob.prototype.addConst = function (newConst) {
        for (var idx_1 = 0; idx_1 < this.consts.length; idx_1++) {
            if (this.consts[idx_1].isEquivalent(newConst)) {
                return idx_1;
            }
        }
        var idx = this.consts.length;
        this.consts.push(newConst);
        return idx;
    };
    return ComponentCompilationJob;
}());
exports.ComponentCompilationJob = ComponentCompilationJob;
var ViewCompilationUnit = (function (_super) {
    __extends(ViewCompilationUnit, _super);
    function ViewCompilationUnit(job, xref, parent) {
        var _this = _super.call(this, xref) || this;
        _this.job = job;
        _this.parent = parent;
        _this.contextVariables = new Map();
        _this.decls = null;
        return _this;
    }
    Object.defineProperty(ViewCompilationUnit.prototype, "compatibility", {
        get: function () {
            return this.job.compatibility;
        },
        enumerable: false,
        configurable: true
    });
    return ViewCompilationUnit;
}(CompilationUnit));
exports.ViewCompilationUnit = ViewCompilationUnit;
