// @componlyco/core/dist/tsc/parse/angular/compiler/src/template/pipeline/src/instruction.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pureFunction = exports.hostProperty = exports.classMapInterpolate = exports.styleMapInterpolate = exports.stylePropInterpolate = exports.attributeInterpolate = exports.propertyInterpolate = exports.textInterpolate = exports.pipeBindV = exports.pipeBind = exports.classMap = exports.styleMap = exports.classProp = exports.styleProp = exports.attribute = exports.property = exports.text = exports.resetView = exports.restoreView = exports.getCurrentView = exports.nextContext = exports.reference = exports.advance = exports.namespaceMath = exports.namespaceSVG = exports.namespaceHTML = exports.pipe = exports.listener = exports.enableBindings = exports.disableBindings = exports.template = exports.elementContainerEnd = exports.elementContainer = exports.elementContainerStart = exports.elementEnd = exports.elementStart = exports.element = void 0;
var o = __importStar(require("../../../output/output_ast"));
var r3_identifiers_1 = require("../../../render3/r3_identifiers");
var ir = __importStar(require("../ir"));
function element(slot, tag, constIndex, localRefIndex, sourceSpan) {
    return elementOrContainerBase(r3_identifiers_1.Identifiers.element, slot, tag, constIndex, localRefIndex, sourceSpan);
}
exports.element = element;
function elementStart(slot, tag, constIndex, localRefIndex, sourceSpan) {
    return elementOrContainerBase(r3_identifiers_1.Identifiers.elementStart, slot, tag, constIndex, localRefIndex, sourceSpan);
}
exports.elementStart = elementStart;
function elementOrContainerBase(instruction, slot, tag, constIndex, localRefIndex, sourceSpan) {
    var args = [o.literal(slot)];
    if (tag !== null) {
        args.push(o.literal(tag));
    }
    if (localRefIndex !== null) {
        args.push(o.literal(constIndex), o.literal(localRefIndex));
    }
    else if (constIndex !== null) {
        args.push(o.literal(constIndex));
    }
    return call(instruction, args, sourceSpan);
}
function elementEnd(sourceSpan) {
    return call(r3_identifiers_1.Identifiers.elementEnd, [], sourceSpan);
}
exports.elementEnd = elementEnd;
function elementContainerStart(slot, constIndex, localRefIndex, sourceSpan) {
    return elementOrContainerBase(r3_identifiers_1.Identifiers.elementContainerStart, slot, null, constIndex, localRefIndex, sourceSpan);
}
exports.elementContainerStart = elementContainerStart;
function elementContainer(slot, constIndex, localRefIndex, sourceSpan) {
    return elementOrContainerBase(r3_identifiers_1.Identifiers.elementContainer, slot, null, constIndex, localRefIndex, sourceSpan);
}
exports.elementContainer = elementContainer;
function elementContainerEnd() {
    return call(r3_identifiers_1.Identifiers.elementContainerEnd, [], null);
}
exports.elementContainerEnd = elementContainerEnd;
function template(slot, templateFnRef, decls, vars, tag, constIndex, sourceSpan) {
    return call(r3_identifiers_1.Identifiers.templateCreate, [
        o.literal(slot),
        templateFnRef,
        o.literal(decls),
        o.literal(vars),
        o.literal(tag),
        o.literal(constIndex),
    ], sourceSpan);
}
exports.template = template;
function disableBindings() {
    return call(r3_identifiers_1.Identifiers.disableBindings, [], null);
}
exports.disableBindings = disableBindings;
function enableBindings() {
    return call(r3_identifiers_1.Identifiers.enableBindings, [], null);
}
exports.enableBindings = enableBindings;
function listener(name, handlerFn) {
    return call(r3_identifiers_1.Identifiers.listener, [
        o.literal(name),
        handlerFn,
    ], null);
}
exports.listener = listener;
function pipe(slot, name) {
    return call(r3_identifiers_1.Identifiers.pipe, [
        o.literal(slot),
        o.literal(name),
    ], null);
}
exports.pipe = pipe;
function namespaceHTML() {
    return call(r3_identifiers_1.Identifiers.namespaceHTML, [], null);
}
exports.namespaceHTML = namespaceHTML;
function namespaceSVG() {
    return call(r3_identifiers_1.Identifiers.namespaceSVG, [], null);
}
exports.namespaceSVG = namespaceSVG;
function namespaceMath() {
    return call(r3_identifiers_1.Identifiers.namespaceMathML, [], null);
}
exports.namespaceMath = namespaceMath;
function advance(delta, sourceSpan) {
    return call(r3_identifiers_1.Identifiers.advance, [
        o.literal(delta),
    ], sourceSpan);
}
exports.advance = advance;
function reference(slot) {
    return o.importExpr(r3_identifiers_1.Identifiers.reference).callFn([
        o.literal(slot),
    ]);
}
exports.reference = reference;
function nextContext(steps) {
    return o.importExpr(r3_identifiers_1.Identifiers.nextContext).callFn(steps === 1 ? [] : [o.literal(steps)]);
}
exports.nextContext = nextContext;
function getCurrentView() {
    return o.importExpr(r3_identifiers_1.Identifiers.getCurrentView).callFn([]);
}
exports.getCurrentView = getCurrentView;
function restoreView(savedView) {
    return o.importExpr(r3_identifiers_1.Identifiers.restoreView).callFn([
        savedView,
    ]);
}
exports.restoreView = restoreView;
function resetView(returnValue) {
    return o.importExpr(r3_identifiers_1.Identifiers.resetView).callFn([
        returnValue,
    ]);
}
exports.resetView = resetView;
function text(slot, initialValue, sourceSpan) {
    var args = [o.literal(slot, null)];
    if (initialValue !== '') {
        args.push(o.literal(initialValue));
    }
    return call(r3_identifiers_1.Identifiers.text, args, sourceSpan);
}
exports.text = text;
function property(name, expression, sanitizer, sourceSpan) {
    var args = [o.literal(name), expression];
    if (sanitizer !== null) {
        args.push(sanitizer);
    }
    return call(r3_identifiers_1.Identifiers.property, args, sourceSpan);
}
exports.property = property;
function attribute(name, expression, sanitizer) {
    var args = [o.literal(name), expression];
    if (sanitizer !== null) {
        args.push(sanitizer);
    }
    return call(r3_identifiers_1.Identifiers.attribute, args, null);
}
exports.attribute = attribute;
function styleProp(name, expression, unit) {
    var args = [o.literal(name), expression];
    if (unit !== null) {
        args.push(o.literal(unit));
    }
    return call(r3_identifiers_1.Identifiers.styleProp, args, null);
}
exports.styleProp = styleProp;
function classProp(name, expression) {
    return call(r3_identifiers_1.Identifiers.classProp, [o.literal(name), expression], null);
}
exports.classProp = classProp;
function styleMap(expression) {
    return call(r3_identifiers_1.Identifiers.styleMap, [expression], null);
}
exports.styleMap = styleMap;
function classMap(expression) {
    return call(r3_identifiers_1.Identifiers.classMap, [expression], null);
}
exports.classMap = classMap;
var PIPE_BINDINGS = [
    r3_identifiers_1.Identifiers.pipeBind1,
    r3_identifiers_1.Identifiers.pipeBind2,
    r3_identifiers_1.Identifiers.pipeBind3,
    r3_identifiers_1.Identifiers.pipeBind4,
];
function pipeBind(slot, varOffset, args) {
    if (args.length < 1 || args.length > PIPE_BINDINGS.length) {
        throw new Error("pipeBind() argument count out of bounds");
    }
    var instruction = PIPE_BINDINGS[args.length - 1];
    return o.importExpr(instruction).callFn(__spreadArray([
        o.literal(slot),
        o.literal(varOffset)
    ], __read(args), false));
}
exports.pipeBind = pipeBind;
function pipeBindV(slot, varOffset, args) {
    return o.importExpr(r3_identifiers_1.Identifiers.pipeBindV).callFn([
        o.literal(slot),
        o.literal(varOffset),
        args,
    ]);
}
exports.pipeBindV = pipeBindV;
function textInterpolate(strings, expressions, sourceSpan) {
    if (strings.length < 1 || expressions.length !== strings.length - 1) {
        throw new Error("AssertionError: expected specific shape of args for strings/expressions in interpolation");
    }
    var interpolationArgs = [];
    if (expressions.length === 1 && strings[0] === '' && strings[1] === '') {
        interpolationArgs.push(expressions[0]);
    }
    else {
        var idx = void 0;
        for (idx = 0; idx < expressions.length; idx++) {
            interpolationArgs.push(o.literal(strings[idx]), expressions[idx]);
        }
        interpolationArgs.push(o.literal(strings[idx]));
    }
    return callVariadicInstruction(TEXT_INTERPOLATE_CONFIG, [], interpolationArgs, [], sourceSpan);
}
exports.textInterpolate = textInterpolate;
function propertyInterpolate(name, strings, expressions, sanitizer, sourceSpan) {
    var interpolationArgs = collateInterpolationArgs(strings, expressions);
    var extraArgs = [];
    if (sanitizer !== null) {
        extraArgs.push(sanitizer);
    }
    return callVariadicInstruction(PROPERTY_INTERPOLATE_CONFIG, [o.literal(name)], interpolationArgs, extraArgs, sourceSpan);
}
exports.propertyInterpolate = propertyInterpolate;
function attributeInterpolate(name, strings, expressions, sanitizer) {
    var interpolationArgs = collateInterpolationArgs(strings, expressions);
    var extraArgs = [];
    if (sanitizer !== null) {
        extraArgs.push(sanitizer);
    }
    return callVariadicInstruction(ATTRIBUTE_INTERPOLATE_CONFIG, [o.literal(name)], interpolationArgs, extraArgs, null);
}
exports.attributeInterpolate = attributeInterpolate;
function stylePropInterpolate(name, strings, expressions, unit) {
    var interpolationArgs = collateInterpolationArgs(strings, expressions);
    var extraArgs = [];
    if (unit !== null) {
        extraArgs.push(o.literal(unit));
    }
    return callVariadicInstruction(STYLE_PROP_INTERPOLATE_CONFIG, [o.literal(name)], interpolationArgs, extraArgs, null);
}
exports.stylePropInterpolate = stylePropInterpolate;
function styleMapInterpolate(strings, expressions) {
    var interpolationArgs = collateInterpolationArgs(strings, expressions);
    return callVariadicInstruction(STYLE_MAP_INTERPOLATE_CONFIG, [], interpolationArgs, [], null);
}
exports.styleMapInterpolate = styleMapInterpolate;
function classMapInterpolate(strings, expressions) {
    var interpolationArgs = collateInterpolationArgs(strings, expressions);
    return callVariadicInstruction(CLASS_MAP_INTERPOLATE_CONFIG, [], interpolationArgs, [], null);
}
exports.classMapInterpolate = classMapInterpolate;
function hostProperty(name, expression) {
    return call(r3_identifiers_1.Identifiers.hostProperty, [o.literal(name), expression], null);
}
exports.hostProperty = hostProperty;
function pureFunction(varOffset, fn, args) {
    return callVariadicInstructionExpr(PURE_FUNCTION_CONFIG, [
        o.literal(varOffset),
        fn,
    ], args, [], null);
}
exports.pureFunction = pureFunction;
function collateInterpolationArgs(strings, expressions) {
    if (strings.length < 1 || expressions.length !== strings.length - 1) {
        throw new Error("AssertionError: expected specific shape of args for strings/expressions in interpolation");
    }
    var interpolationArgs = [];
    if (expressions.length === 1 && strings[0] === '' && strings[1] === '') {
        interpolationArgs.push(expressions[0]);
    }
    else {
        var idx = void 0;
        for (idx = 0; idx < expressions.length; idx++) {
            interpolationArgs.push(o.literal(strings[idx]), expressions[idx]);
        }
        interpolationArgs.push(o.literal(strings[idx]));
    }
    return interpolationArgs;
}
function call(instruction, args, sourceSpan) {
    var expr = o.importExpr(instruction).callFn(args, sourceSpan);
    return ir.createStatementOp(new o.ExpressionStatement(expr, sourceSpan));
}
var TEXT_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.textInterpolate,
        r3_identifiers_1.Identifiers.textInterpolate1,
        r3_identifiers_1.Identifiers.textInterpolate2,
        r3_identifiers_1.Identifiers.textInterpolate3,
        r3_identifiers_1.Identifiers.textInterpolate4,
        r3_identifiers_1.Identifiers.textInterpolate5,
        r3_identifiers_1.Identifiers.textInterpolate6,
        r3_identifiers_1.Identifiers.textInterpolate7,
        r3_identifiers_1.Identifiers.textInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.textInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var PROPERTY_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.propertyInterpolate,
        r3_identifiers_1.Identifiers.propertyInterpolate1,
        r3_identifiers_1.Identifiers.propertyInterpolate2,
        r3_identifiers_1.Identifiers.propertyInterpolate3,
        r3_identifiers_1.Identifiers.propertyInterpolate4,
        r3_identifiers_1.Identifiers.propertyInterpolate5,
        r3_identifiers_1.Identifiers.propertyInterpolate6,
        r3_identifiers_1.Identifiers.propertyInterpolate7,
        r3_identifiers_1.Identifiers.propertyInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.propertyInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var STYLE_PROP_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.styleProp,
        r3_identifiers_1.Identifiers.stylePropInterpolate1,
        r3_identifiers_1.Identifiers.stylePropInterpolate2,
        r3_identifiers_1.Identifiers.stylePropInterpolate3,
        r3_identifiers_1.Identifiers.stylePropInterpolate4,
        r3_identifiers_1.Identifiers.stylePropInterpolate5,
        r3_identifiers_1.Identifiers.stylePropInterpolate6,
        r3_identifiers_1.Identifiers.stylePropInterpolate7,
        r3_identifiers_1.Identifiers.stylePropInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.stylePropInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var ATTRIBUTE_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.attribute,
        r3_identifiers_1.Identifiers.attributeInterpolate1,
        r3_identifiers_1.Identifiers.attributeInterpolate2,
        r3_identifiers_1.Identifiers.attributeInterpolate3,
        r3_identifiers_1.Identifiers.attributeInterpolate4,
        r3_identifiers_1.Identifiers.attributeInterpolate5,
        r3_identifiers_1.Identifiers.attributeInterpolate6,
        r3_identifiers_1.Identifiers.attributeInterpolate7,
        r3_identifiers_1.Identifiers.attributeInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.attributeInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var STYLE_MAP_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.styleMap,
        r3_identifiers_1.Identifiers.styleMapInterpolate1,
        r3_identifiers_1.Identifiers.styleMapInterpolate2,
        r3_identifiers_1.Identifiers.styleMapInterpolate3,
        r3_identifiers_1.Identifiers.styleMapInterpolate4,
        r3_identifiers_1.Identifiers.styleMapInterpolate5,
        r3_identifiers_1.Identifiers.styleMapInterpolate6,
        r3_identifiers_1.Identifiers.styleMapInterpolate7,
        r3_identifiers_1.Identifiers.styleMapInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.styleMapInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var CLASS_MAP_INTERPOLATE_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.classMap,
        r3_identifiers_1.Identifiers.classMapInterpolate1,
        r3_identifiers_1.Identifiers.classMapInterpolate2,
        r3_identifiers_1.Identifiers.classMapInterpolate3,
        r3_identifiers_1.Identifiers.classMapInterpolate4,
        r3_identifiers_1.Identifiers.classMapInterpolate5,
        r3_identifiers_1.Identifiers.classMapInterpolate6,
        r3_identifiers_1.Identifiers.classMapInterpolate7,
        r3_identifiers_1.Identifiers.classMapInterpolate8,
    ],
    variable: r3_identifiers_1.Identifiers.classMapInterpolateV,
    mapping: function (n) {
        if (n % 2 === 0) {
            throw new Error("Expected odd number of arguments");
        }
        return (n - 1) / 2;
    },
};
var PURE_FUNCTION_CONFIG = {
    constant: [
        r3_identifiers_1.Identifiers.pureFunction0,
        r3_identifiers_1.Identifiers.pureFunction1,
        r3_identifiers_1.Identifiers.pureFunction2,
        r3_identifiers_1.Identifiers.pureFunction3,
        r3_identifiers_1.Identifiers.pureFunction4,
        r3_identifiers_1.Identifiers.pureFunction5,
        r3_identifiers_1.Identifiers.pureFunction6,
        r3_identifiers_1.Identifiers.pureFunction7,
        r3_identifiers_1.Identifiers.pureFunction8,
    ],
    variable: r3_identifiers_1.Identifiers.pureFunctionV,
    mapping: function (n) { return n; },
};
function callVariadicInstructionExpr(config, baseArgs, interpolationArgs, extraArgs, sourceSpan) {
    var n = config.mapping(interpolationArgs.length);
    if (n < config.constant.length) {
        return o.importExpr(config.constant[n])
            .callFn(__spreadArray(__spreadArray(__spreadArray([], __read(baseArgs), false), __read(interpolationArgs), false), __read(extraArgs), false), sourceSpan);
    }
    else if (config.variable !== null) {
        return o.importExpr(config.variable)
            .callFn(__spreadArray(__spreadArray(__spreadArray([], __read(baseArgs), false), [o.literalArr(interpolationArgs)], false), __read(extraArgs), false), sourceSpan);
    }
    else {
        throw new Error("AssertionError: unable to call variadic function");
    }
}
function callVariadicInstruction(config, baseArgs, interpolationArgs, extraArgs, sourceSpan) {
    return ir.createStatementOp(callVariadicInstructionExpr(config, baseArgs, interpolationArgs, extraArgs, sourceSpan)
        .toStmt());
}
