// @componlyco/berkoukes/dist/tsc/components/PageHeader/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Settings } from "lucide-react";
import { ButtonType } from "../Button";
import { PageHeader } from "./index";
export default {
    title: "PageHeader",
    component: PageHeader,
    args: {
        link: "https://www.google.com/Link1/Link2/Link3/",
        buttons: [
            { type: ButtonType.Borders, label: "Button1", onClick: function () { return console.log("Button1 clicked"); } },
            {
                type: ButtonType.Secondary,
                label: "Button2",
                onClick: function () { return console.log("Button2 clicked"); },
            },
            { type: ButtonType.Primary, label: "Button3", onClick: function () { return console.log("Button3 clicked"); } },
            {
                type: ButtonType.NoBorders,
                icon: _jsx(Settings, {}),
                onClick: function () { return console.log("Gear Icon clicked"); },
            },
        ],
        breadcrumbProps: {
            onClickHome: function () { return alert('go home'); },
            items: [
                {
                    label: 'link 1',
                    onClick: function () { return alert('go to link 1'); }
                },
                {
                    label: 'link 2',
                    onClick: function () { return alert('go to link 2'); }
                }
            ]
        },
    },
};
var Template = function (args) { return _jsx(PageHeader, __assign({}, args)); };
export var Default = Template.bind({});
export var Header1 = Template.bind({});
Header1.args = {
    backgroundImage: "rgb(255,229,166)",
    avatarImage: "https://img.freepik.com/free-psd/3d-illustration-person-with-sunglasses_23-2149436188.jpg?w=2000",
    title: "Jacob Mills",
    subtitle: "Strategic Product Designer for corporate ventures, startups & scale ups",
};
export var Header2 = Template.bind({});
Header2.args = {
    backgroundImage: "linear-gradient(153deg, rgba(244, 244, 244, 1) 0%, rgba(201, 203, 212, 1) 76%, rgba(244, 244, 244, 1) 100%)",
    avatarImage: "https://img.freepik.com/premium-psd/3d-chef-cartoon-character-avatar-isolated-3d-rendering_235528-1014.jpg?w=740",
    title: "Monica Gellar",
    subtitle: "A chef known for her cleanliness, competitiveness and obsessive-compulsive nature",
};
