// @componlyco/berkoukes/dist/tsc/components/Tabs/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { TabType, Tabs } from "./index";
export default {
    title: "Tabs",
    component: Tabs,
    args: {
        value: "My Details 1",
        tabItems: [
            {
                text: "My Details 1",
                current: true,
            },
            {
                text: "My Details 2",
            },
            {
                text: "My Details 3",
            },
        ],
    },
};
var Template = function (args) {
    return _jsx(Tabs, __assign({}, args));
};
export var Primary = Template.bind({});
Primary.args = {
    type: TabType.Primary,
};
export var Underline = Template.bind({});
Underline.args = {
    type: TabType.Underline,
};
export var FullWidthPrimary = Template.bind({});
FullWidthPrimary.args = {
    fullWidth: true,
    sx: {
        width: 'calc(100vw - 32px)'
    },
    type: TabType.Primary,
};
export var FullWidthUnderline = Template.bind({});
FullWidthUnderline.args = {
    fullWidth: true,
    sx: {
        width: 'calc(100vw - 32px)'
    },
    type: TabType.Underline,
};
