// @componlyco/berkoukes/dist/tsc/components/InputParagraph/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { InputParagraph } from "./index";
export default {
    title: "InputParagraph",
    component: InputParagraph,
    args: {
        label: "Label",
        placeholder: "Enter text here...",
        optionalText: "Optional",
        helperText: "Helper text",
    },
};
var Template = function (args) { return _jsx(InputParagraph, __assign({}, args)); };
export var Default = Template.bind({});
export var Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
};
export var Error = Template.bind({});
Error.args = {
    error: "Helper Text",
};
export var WithoutOptional = Template.bind({});
WithoutOptional.args = {
    optionalText: undefined,
};
