// @componlyco/berkoukes/dist/tsc/design-tokens.js
export var Colors = {
    DarkGrey: {
        1300: "#040404",
        900: "#101828",
        600: "#475467",
        200: "#667181",
        75: "#B4B9C1",
        50: "#F9FAFB",
        black: "#000000",
        cf0: "#EAECF0",
    },
    Neutral: {
        1300: "#0A090B",
        1100: "#2D2B32",
        1000: "#4F4D55",
        900: "#212121",
        800: "#404040",
        700: "#616161",
        600: "#DCDCDE",
        500: "#9E9E9E",
        50: "#EDEEF0",
        25: "#F5F6F8",
        0: "#E6E6E6",
        white: "#ffffff",
        ece: "#ECECEC",
        grey: "#615E83",
    },
    Border: {
        filled: "#D0D5DD",
    },
    Primary: {
        green: "#6CE9A6",
        red: "#FF7262",
        blue: "#007BFF",
    },
    Default: {
        Background: "#007BFF",
        Text: "#FFFFFF",
        Border: "#9E9E9E",
        Avatar: "#404040",
    },
    Hovered: {
        Background: "#F5F6F8",
        Text: "#FFFFFF",
        Border: "#007BFF",
    },
    Focused: {
        Background: "linear-gradient(180deg, #4C4C4C 0%, #000000 100%)",
        Text: "#FFFFFF",
        Border: "#0000001A",
        Fade: "#E6F2FF",
    },
    Disabled: {
        Background: "#ECEEF1",
        Text: "#9E9E9E",
        Gray: "#F5F6F8",
    },
    Breadcrumb: "#475467",
    Badge: {
        neutral: {
            Label: "#000000",
            Background: {
                outline: "#ffffff",
                filled: "#ffffff",
            },
            Border: {
                filled: "#D0D5DD",
                outline: "#000000",
            },
        },
        accent: {
            Label: "#007BFF",
            Background: {
                outline: "#ffffff",
                filled: "#E6F2FF",
            },
            Border: {
                filled: "none",
                outline: "#007BFF",
            },
        },
        error: {
            Label: "#E01E27",
            Background: {
                outline: "#ffffff",
                filled: "#FCE9E9",
            },
            Border: {
                filled: "none",
                outline: "#E01E27",
            },
        },
        warning: {
            Label: "#EB972A",
            Background: {
                outline: "#ffffff",
                filled: "#FDF5EA",
            },
            Border: {
                filled: "none",
                outline: "#EB972A",
            },
        },
        success: {
            Label: "#139841",
            Background: {
                outline: "#ffffff",
                filled: "#E7F5EC",
            },
            Border: {
                filled: "none",
                outline: "#139841",
            },
        },
    },
    ProgressIndicator: {
        Step: {
            Default: {
                Background: "#F5F6F8",
                Text: "#616161",
            },
            Active: {
                Background: "#007BFF",
                Text: "#E6F2FF",
            },
            Completed: {
                Background: "#007BFF",
                Text: "#FFFFFF",
            },
        },
        VerticalSeparator: {
            Default: {
                Border: "#D0D5DD",
            },
            Active: {
                Border: "#007BFF",
            },
        },
    },
    Button: {
        primary: {
            Default: {
                Text: "#FFFFFF",
                Background: "linear-gradient(180deg, #4C4C4C 0%, #000000 100%)",
                Border: "var(--main-neutral-1000-black, #000000)",
            },
            Hover: {
                Text: "#000000",
                Background: "#F5F6F8",
                Border: "#0000001A",
            },
            Focus: {
                Text: "#FFFFFF",
                Background: "linear-gradient(180deg, #4C4C4C 0%, #000000 100%)",
                Border: "var(--main-neutral-1000-black, #000000)",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        secondary: {
            Default: {
                Text: "#FFFFFF",
                Background: "#000000",
                Border: "none",
            },
            Hover: {
                Text: "#000000",
                Background: "#F5F6F8",
                Border: "none",
            },
            Focus: {
                Text: "#FFFFFF",
                Background: "#000000",
                Border: "none",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        tab: {
            Default: {
                Background: "#E6F2FF",
                Text: "#007BFF",
                Border: "none",
            },
            Hover: {
                Background: "#E6F2FF",
                Text: "#007BFF",
                Border: "none",
            },
            Focus: {
                Background: "#E6F2FF",
                Text: "#007BFF",
                Border: "none",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        borders: {
            Default: {
                Text: "#000000",
                Background: "#FFFFFF",
                Border: "#D0D5DD",
            },
            Hover: {
                Text: "#000000",
                Background: "#F5F6F8",
                Border: "#D0D5DD",
            },
            Focus: {
                Text: "#000000",
                Background: "#FFFFFF",
                Border: "#D0D5DD",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        "no-borders": {
            Default: {
                Text: "#212121",
                Background: "#FFFFFF",
                Border: "none",
            },
            Hover: {
                Text: "#000000",
                Background: "#F5F6F8",
                Border: "none",
            },
            Focus: {
                Text: "#212121",
                Background: "#FFFFFF",
                Border: "none",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#FFFFFF",
                Border: "none",
            },
        },
        destructive: {
            Default: {
                Text: "#E01E27",
                Background: "#FCE9E9",
                Border: "none",
            },
            Hover: {
                Text: "#E01E27",
                Background: "#F2A3A6",
                Border: "none",
            },
            Focus: {
                Text: "#E01E27",
                Background: "#FCE9E9",
                Border: "none",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        success: {
            Default: {
                Text: "#139841",
                Background: "#E7F5EC",
                Border: "none",
            },
            Hover: {
                Text: "#139841",
                Background: "#9ED5B1",
                Border: "none",
            },
            Focus: {
                Text: "#139841",
                Background: "#E7F5EC",
                Border: "none",
            },
            Disabled: {
                Text: "#C3C5C8",
                Background: "#F5F6F8",
                Border: "none",
            },
        },
        "text-primary": {
            Default: {
                Text: "#007BFF",
                Background: "#FFFFFF",
                Border: "none",
            },
            Hover: {
                Text: "#0056B3",
                Background: "#FFFFFF",
                Border: "none",
            },
            Focus: {
                Text: "#000000",
                Background: "#FFFFFF",
                Border: "#007BFF",
            },
            Disabled: {
                Text: "#D0D5DD",
                Background: "#FFFFFF",
                Border: "none",
            },
        },
        "text-secondary": {
            Default: {
                Text: "#616161",
                Background: "#FFFFFF",
                Border: "none",
            },
            Hover: {
                Text: "#000000",
                Background: "#FFFFFF",
                Border: "none",
            },
            Focus: {
                Text: "#475467",
                Background: "#FFFFFF",
                Border: "#007BFF",
            },
            Disabled: {
                Text: "#D0D5DD",
                Background: "#FFFFFF",
                Border: "none",
            },
        },
        "text-destructive": {
            Default: {
                Text: "#E01E27",
                Background: "#FFFFFF",
                Border: "none",
            },
            Focus: {
                Text: "#E01E27",
                Background: "#FFFFFF",
                Border: "none",
            },
            Hover: {
                Text: "#9D151B",
                Background: "#FFFFFF",
                Border: "none",
            },
            Disabled: {
                Text: "#D0D5DD",
                Background: "#FFFFFF",
                Border: "none",
            },
        },
        warning: {
            Default: {
                Text: "#EB972A",
                Background: "#FDF5EA",
                Border: "none",
            },
            Focus: {
                Text: "#EB972A",
                Background: "#FDF5EA",
                Border: "none",
            },
            Hover: {
                Text: "#EB974F",
                Background: "#F7D4A8",
                Border: "none",
            },
            Disabled: {
                Text: "#EB972A",
                Background: "#FDF5EA",
                Border: "none",
            },
        },
    },
    UploadItem: {
        Close: "#7F7D83",
        Save: "#1964FF",
    },
    Tabs: {
        Background: "#E6F2FF",
        Text: "#007BFF",
    },
    UploadContainer: {
        Text: "#475467",
        Hover: {
            Border: "#007BFF",
            Background: "#E6F2FF",
        },
        Default: {
            Border: "#949CA7",
            Background: "#E6F2FF",
        },
        Disabled: {
            Border: "#949CA7",
            Background: "#FFFFFF",
        },
        Icon: {
            Text: "#404040",
            Background: "#F5F6F8",
        },
    },
    Card: {
        Default: {
            Border: "2px solid #F5F6F8",
            Background: "#E6F2FF",
        },
        Text: {
            Default: {
                Text: "#475467",
            },
        },
        Hover: {
            Border: "2px solid #007BFF",
            PrimaryTextColor: "#007BFF",
        },
        Focus: "0px 0px 0px 6px rgba(0, 123, 255, 0.15)",
    },
    DatePicker: {
        WeekLabel: "#000000",
        HeaderLabel: "#212121",
    },
    Icons: {
        Twitter: {
            Color: "#1DA1F2",
        },
        Linkedin: {
            Color: "#FFFFFF",
            Background: "#0A66C2",
        },
    },
};
// eslint-disable-next-line import/no-unused-modules
export var Sizes = {
    Checkbox: {
        Small: "24px",
        Normal: "32px",
    },
    Toggle: {
        Track: {
            Width: "44px",
            Height: "32px",
        },
        Thumb: {
            Width: "20px",
            Height: "20px",
        },
    },
    InputParagraph: {
        Radius: "16px",
        ErrorHeight: "14px",
        LabelHeight: "24px",
    },
    Avatar: {
        xxs: "16px",
        xs: "24px",
        s: "32px",
        m: "40px",
        l: "48px",
        xl: "56px",
        xxl: "64px",
        profile: "160px",
    },
    Button: {
        Border: {
            primary: {
                Default: "2px",
                Focus: "2px",
            },
            secondary: {
                Default: "0",
                Focus: "2px",
            },
            tab: {
                Default: "0",
                Focus: "2px",
            },
            borders: {
                Default: "1px",
                Focus: "2px",
            },
            "no-borders": {
                Default: "0",
                Focus: "2px",
            },
            destructive: {
                Default: "0",
                Focus: "2px",
            },
            success: {
                Default: "0",
                Focus: "2px",
            },
            "text-primary": {
                Default: "0",
                Focus: "2px",
            },
            "text-secondary": {
                Default: "0",
                Focus: "2px",
            },
            "text-destructive": {
                Default: "0",
                Focus: "2px",
            },
            warning: {
                Default: "0",
                Focus: "2px",
            },
        },
    },
};
export var Spacing = {
    Button: {
        primary: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        secondary: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        tab: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        borders: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        "no-borders": {
            Default: "0 24px",
            Small: "6px 20px",
        },
        destructive: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        success: {
            Default: "0 24px",
            Small: "6px 20px",
        },
        "text-primary": {
            Default: "2px",
            Small: "2px",
        },
        "text-secondary": {
            Default: "2px",
            Small: "2px",
        },
        "text-destructive": {
            Default: "0 24px",
            Small: "0 20px",
        },
        warning: {
            Default: "6px 24px",
            Small: "6px 20px",
        },
    },
};
export var FontSizes = {
    Title: {
        fontSize: "56px",
        fontWeight: 400,
        lineHeight: "64px",
    },
    Title2: {
        fontSize: "40px",
        fontWeight: 700,
        lineHeight: "48px",
    },
    Subtitle: {
        fontSize: "32px",
        fontWeight: 400,
        lineHeight: "24px",
    },
    Subtitle2: {
        fontSize: "24px",
        fontWeight: 700,
        lineHeight: "24px",
    },
    Subtitle3: {
        fontSize: "14px",
        fontWeight: 700,
        lineHeight: "24px",
    },
    Body1: {
        fontSize: "18px",
        fontWeight: 500,
        lineHeight: "24px",
    },
    Body2: {
        fontSize: "16px",
        fontWeight: 400,
        lineHeight: "24px",
    },
    Body3: {
        fontSize: "14px",
        fontWeight: 400,
        lineHeight: "24px",
    },
    Button: {
        fontSize: "16px",
        fontWeight: 700,
        lineHeight: "24px",
    },
    Link: {
        fontSize: "16px",
        fontWeight: 700,
        lineHeight: "16px",
    },
    Helper: {
        fontSize: "12px",
        fontWeight: 400,
        lineHeight: "16px",
    },
    Upload: {
        Header: "24px",
        Body: "16px",
    },
};
export var Layout = {
    xxs: "2px",
    xs: "4px",
    sm: "8px",
    md: "12px",
    lg: "16px",
    TBR: "20px",
    xl: "24px",
    "2xl": "32px",
    "3xl": "48px",
    "4xl": "64px",
    "5xl": "96px",
};
export var BoxShadows = {
    Button: {
        Focus: {
            primary: "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
            secondary: "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
            tab: "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
            borders: "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
            "no-borders": "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
            destructive: "0px 0px 0px 3px rgba(221, 46, 46, 0.1)",
            success: "0px 0px 0px 3px #9ED5B1",
            "text-primary": "0px 0px 0px 3px rgba(86, 97, 129, 0.08)",
            "text-secondary": "0px 0px 0px 3px rgba(86, 97, 129, 0.15)",
            "text-destructive": "0px 0px 0px 3px rgba(221, 46, 46, 0.1)",
            warning: "0pz 0pz 0pz 3px #EB972A",
        },
    },
    VerificationInput: "0px 2px 14px -1px rgba(10, 9, 11, 0.09), 0px 1px 2px -1px rgba(10, 9, 11, 0.04)",
    Input: "0px 0px 0px 3px rgba(36, 100, 241, 0.15)",
    InputParagraph: {
        Error: "0px 0px 0px 3px rgba(221, 46, 46, 0.1)",
    },
    UserProfile: {
        GroupActive: "0px 0px 0px 3px rgba(0, 123, 255, 0.15)",
        MenuShadow: "0px 2px 14px -1px rgba(10, 9, 11, 0.09), 0px 1px 2px -1px rgba(10, 9, 11, 0.04)",
        DropdownShadow: "0px 18px 36px -5px rgba(10, 9, 11, 0.16), 0px 26px 41px -23px rgba(10, 9, 11, 0.12)",
    },
};
export var Scrollbar = function (width) {
    if (width === void 0) { width = 5; }
    return ({
        "&::-webkit-scrollbar": {
            width: width,
        },
        /* Track */
        "&::-webkit-scrollbar-track": {
            background: "transparent",
        },
        /* Handle */
        "&::-webkit-scrollbar-thumb": {
            background: "#888",
        },
        /* Handle on hover */
        "&::-webkit-scrollbar-thumb:hover": {
            background: "#555",
        },
    });
};
