// @componlyco/berkoukes/dist/tsc/components/ChartCard/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Settings } from "lucide-react";
import { Checkbox } from "../Checkbox";
import { ChartCard } from "./index";
export default {
    title: "ChartCard",
    component: ChartCard,
    args: {
        tabsValue: "Colors",
        tabsItems: [
            { text: "All" },
            { text: "Colors" },
            { text: "Gradients" },
            { text: "Typos" },
            { text: "Spacers" },
            { text: "Opacity" },
            { text: "Stroke" },
        ],
        tags: [
            {
                icon: [_jsx(Checkbox, { size: "small" }), _jsx(Settings, { fill: "red", color: "white" })],
                label: "Nietzsche",
            },
            {
                icon: [_jsx(Checkbox, { size: "small" }), _jsx(Settings, { fill: "red", color: "white" })],
                label: "Nietzsche",
            },
            {
                icon: [_jsx(Checkbox, { size: "small" }), _jsx(Settings, { fill: "red", color: "white" })],
                label: "Nietzsche",
            },
        ]
    },
};
var Template = function (args) { return _jsx(ChartCard, __assign({}, args)); };
export var Default = Template.bind({});
Default.args = {
    primaryText: "Name of the card",
    secondaryText: "This graph displays the status of components over time, indicating whether they are from the DS, deprecated or custom.",
    buttonText: "Full report",
};
export var Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
    primaryText: "Name of the card",
    secondaryText: "This graph displays the status of components over time, indicating whether they are from the DS, deprecated or custom.",
    buttonText: "Full report"
};
