// @componlyco/berkoukes/dist/tsc/components/TextInput/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Search as SearchIcon } from "lucide-react";
import { TextInput } from "./index";
export default {
    title: "TextInput",
    component: TextInput,
    args: {
        label: "Label",
        placeholder: "Enter text",
        type: "text",
        optionalText: "Optional",
        helperText: "Helper Text",
        helpTooltipProps: {
            position: "top",
            title: "Header",
            description: "In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content.",
        }
    },
};
var Template = function (args) { return _jsx(TextInput, __assign({}, args)); };
export var Default = Template.bind({});
export var Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
};
export var Error = Template.bind({});
Error.args = {
    error: "Helper Text",
};
export var WithoutOptional = Template.bind({});
WithoutOptional.args = {
    optionalText: undefined,
};
export var Search = Template.bind({});
Search.args = {
    label: "Search",
    type: "search",
    placeholder: "search",
    icon: _jsx(SearchIcon, {}),
};
