// @componlyco/core/dist/tsc/parse/angular/compiler/src/ml_parser/xml_tags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getXmlTagDefinition = exports.XmlTagDefinition = void 0;
var tags_1 = require("./tags");
var XmlTagDefinition = (function () {
    function XmlTagDefinition() {
        this.closedByParent = false;
        this.implicitNamespacePrefix = null;
        this.isVoid = false;
        this.ignoreFirstLf = false;
        this.canSelfClose = true;
        this.preventNamespaceInheritance = false;
    }
    XmlTagDefinition.prototype.requireExtraParent = function (currentParent) {
        return false;
    };
    XmlTagDefinition.prototype.isClosedByChild = function (name) {
        return false;
    };
    XmlTagDefinition.prototype.getContentType = function () {
        return tags_1.TagContentType.PARSABLE_DATA;
    };
    return XmlTagDefinition;
}());
exports.XmlTagDefinition = XmlTagDefinition;
var _TAG_DEFINITION = new XmlTagDefinition();
function getXmlTagDefinition(tagName) {
    return _TAG_DEFINITION;
}
exports.getXmlTagDefinition = getXmlTagDefinition;
