// @componlyco/berkoukes/dist/tsc/components/TagsUsageCard/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { TagsUsageCard } from "./index";
export default {
    title: "TagsUsageCard",
    component: TagsUsageCard,
    args: {
        maxHeight: 260,
        sx: {
            minWidth: 350,
        },
        paginationProps: {
            value: 2,
            total: 8,
            onChange: function () { return null; },
            withControls: false,
        },
    },
};
var Template = function (args) { return (_jsx(TagsUsageCard, __assign({}, args))); };
export var Default = Template.bind({});
Default.args = {
    type: "visibility",
    trend: "up",
    title: "Increasing tags",
    items: [
        {
            title: "alugar",
            count: "87",
        },
        {
            title: "alugar",
            count: "60",
        },
        {
            title: "alugar",
            count: "57",
        },
        {
            title: "alugar",
            count: "34",
        },
        {
            title: "alugar",
            count: "25",
        },
        {
            title: "alugar",
            count: "14",
        },
        {
            title: "alugar",
            count: "8",
        },
    ],
};
export var Usage = Template.bind({});
Usage.args = {
    type: "usage",
    title: "Tags by Usage",
    items: [
        {
            title: "alugar",
            count: "1 345",
        },
        {
            title: "alugar",
            count: "678",
        },
        {
            title: "alugar",
            count: "451",
        },
        {
            title: "alugar",
            count: "356",
        },
        {
            title: "alugar",
            count: "245",
        },
        {
            title: "alugar",
            count: "218",
        },
        {
            title: "alugar",
            count: "112",
        },
    ],
};
