// @componlyco/berkoukes/dist/tsc/components/SecondaryMobileHeader/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { X } from "lucide-react";
import { ButtonType } from "../Button";
import SecondaryHeaderStory from "../SecondaryHeader/index.stories";
import { SecondaryMobileHeader } from "./index";
export default {
    title: "SecondaryMobileHeader",
    component: SecondaryMobileHeader,
    args: __assign(__assign({}, SecondaryHeaderStory.args), { endButtons: [
            {
                type: ButtonType.Borders,
                icon: _jsx(X, {}),
            },
        ] }),
};
export var Template = function (args) { return (_jsx("div", __assign({ style: { width: "90vw" } }, { children: _jsx(SecondaryMobileHeader, __assign({}, args)) }))); };
