// @componlyco/berkoukes/dist/tsc/components/Header/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Hexagon } from "lucide-react";
import { ButtonSizes, ButtonType } from "../Button";
import { Header, HeaderType } from "./index";
export default {
    title: "Header",
    component: Header,
    args: {
        withDivider: true,
        tabs: [
            {
                type: ButtonType.Secondary,
                size: ButtonSizes.Normal,
                label: "Dashboard",
            },
            {
                type: ButtonType.NoBorders,
                size: ButtonSizes.Normal,
                label: "Brands",
            },
            {
                type: ButtonType.NoBorders,
                size: ButtonSizes.Normal,
                label: "Projects",
            },
            {
                type: ButtonType.NoBorders,
                size: ButtonSizes.Normal,
                label: "Books",
            },
            {
                type: ButtonType.NoBorders,
                size: ButtonSizes.Normal,
                label: "Design Systems",
            },
        ],
        profile: {
            src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
            name: "Team name",
            handle: "@teamhandle",
            avatarType: "image",
            width: "250px",
        },
    },
};
var Template = function (args) { return _jsx(Header, __assign({}, args)); };
export var App = Template.bind({});
App.args = {
    type: HeaderType.App,
    endButtons: [
        {
            type: ButtonType.Primary,
            size: ButtonSizes.Normal,
            icon: _jsx(Hexagon, {}),
        },
    ],
    profile: {
        src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
        name: "Team name",
        handle: "@teamhandle",
        avatarType: "image",
        width: "250px",
    },
};
export var ActionHeader = Template.bind({});
ActionHeader.args = {
    type: HeaderType.Action,
    tabs: [
        {
            type: ButtonType.Primary,
            size: ButtonSizes.Normal,
            icon: _jsx(Hexagon, {}),
        },
        {
            type: ButtonType.Primary,
            size: ButtonSizes.Normal,
            icon: _jsx(Hexagon, {}),
        },
        {
            type: ButtonType.NoBorders,
            size: ButtonSizes.Normal,
            label: "Product",
        },
        {
            type: ButtonType.NoBorders,
            size: ButtonSizes.Normal,
            label: "Product",
        },
        {
            type: ButtonType.NoBorders,
            size: ButtonSizes.Normal,
            label: "Product",
        },
        {
            type: ButtonType.NoBorders,
            size: ButtonSizes.Normal,
            label: "Product",
        },
    ],
    startButtons: [
        {
            type: ButtonType.Borders,
            size: ButtonSizes.Normal,
            label: "Back",
        },
    ],
    endButtons: [
        {
            type: ButtonType.Borders,
            size: ButtonSizes.Normal,
            label: "Back",
        },
        {
            type: ButtonType.Borders,
            size: ButtonSizes.Normal,
            label: "Back",
        },
        {
            type: ButtonType.Primary,
            size: ButtonSizes.Normal,
            icon: _jsx(Hexagon, {}),
        },
    ],
    profile: {
        src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
        name: "Team name",
        handle: "@teamhandle",
        avatarType: "image",
        width: "250px",
    },
};
