// @componlyco/core/dist/tsc/scan/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getComponentStoryURLs = exports.extractStorybookURLs = void 0;
var path_1 = require("path");
var storybook_parser_1 = require("@componlyco/storybook-parser");
var chalk_1 = __importDefault(require("chalk"));
var lodash_1 = require("lodash");
function extractStorybookURLs(_a) {
    var storybookBuild = _a.storybookBuild, baseURL = _a.baseURL, clientName = _a.clientName, staticDomain = _a.staticDomain;
    return __awaiter(this, void 0, void 0, function () {
        var storybookBuildFolder, stories, err_1;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    storybookBuildFolder = storybookBuild !== null && storybookBuild !== void 0 ? storybookBuild : (0, path_1.join)((process.cwd(), "storybook-static"));
                    stories = {};
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 4, , 5]);
                    return [4, (0, storybook_parser_1.extractStories)(storybookBuildFolder)];
                case 2:
                    _b.sent();
                    return [4, (0, storybook_parser_1.getStoryURLs)({
                            storybookBuildFolder: storybookBuildFolder,
                            baseURL: baseURL,
                            clientName: clientName,
                            staticDomain: staticDomain,
                        })];
                case 3:
                    stories = _b.sent();
                    return [3, 5];
                case 4:
                    err_1 = _b.sent();
                    if (err_1.code === "MODULE_NOT_FOUND") {
                        console.log(chalk_1.default.red("storybook-static directory is not present."));
                    }
                    return [3, 5];
                case 5: return [2, stories];
            }
        });
    });
}
exports.extractStorybookURLs = extractStorybookURLs;
function filterStoriesByFilePath(stories, filePath) {
    var filteredStories = stories.filter(function (_a) {
        var importPath = _a.importPath;
        var posixImportPath = importPath.split(path_1.sep).join(path_1.posix.sep);
        return posixImportPath.includes(filePath);
    });
    return filteredStories;
}
function getFilteredStoryURLs(stories, urlDictionary) {
    var urls = stories.flatMap(function (_a) {
        var _b;
        var id = _a.id;
        var urlID = id.split("--")[0];
        return (_b = urlDictionary[urlID]) !== null && _b !== void 0 ? _b : [];
    });
    return urls;
}
function getComponentStoryURLs(component, storyURLOutput) {
    var _a, _b;
    var storyFilePaths = (_a = component.stories) !== null && _a !== void 0 ? _a : [];
    var meta = storyURLOutput.meta, urls = storyURLOutput.urls;
    var stories = Object.values((_b = meta === null || meta === void 0 ? void 0 : meta.stories) !== null && _b !== void 0 ? _b : {});
    var storybookURLs = storyFilePaths.flatMap(function (filePath) {
        var filteredStories = filterStoriesByFilePath(stories, filePath);
        var filteredURLs = getFilteredStoryURLs(filteredStories, urls !== null && urls !== void 0 ? urls : {});
        return filteredURLs;
    });
    return (0, lodash_1.uniq)(storybookURLs);
}
exports.getComponentStoryURLs = getComponentStoryURLs;
