// @componlyco/berkoukes/dist/tsc/components/UserProfile/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { SettingsIcon } from "lucide-react";
import { UserDropdownTypes, UserProfile } from "./index";
export default {
    title: "UserProfile",
    component: UserProfile,
    args: {
        src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80",
        name: "User Name",
        handle: "@username",
        width: "249px",
        items: [
            { label: "Resources", onClick: function () { return alert("Resources"); } },
            { label: "Billing" },
            { label: "Settings" },
            { label: "Logout", icon: _jsx(SettingsIcon, {}) },
        ],
    },
};
var Template = function (args) { return _jsx(UserProfile, __assign({}, args)); };
export var Contained = Template.bind({});
Contained.args = {
    type: UserDropdownTypes.Contained,
    avatarSize: "s",
    avatarType: "image",
};
export var NotContained = Template.bind({});
NotContained.args = {
    avatarSize: "s",
    avatarType: "image",
};
