// @componlyco/core/dist/tsc/parse/stencil/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseStencil = exports.createProject = exports.readFileNames = exports.getProjectInstance = void 0;
var path_1 = require("path");
var promises_1 = require("fs/promises");
var ts_morph_1 = require("ts-morph");
var minimatch_1 = require("minimatch");
var names_1 = require("../react/utils/names");
function getProjectInstance() {
    var project = new ts_morph_1.Project({
        compilerOptions: {
            target: ts_morph_1.ScriptTarget.Latest,
        },
    });
    return project;
}
exports.getProjectInstance = getProjectInstance;
function readFileNames(dir, ignorePatterns) {
    return __awaiter(this, void 0, void 0, function () {
        var dirents, updatedIgnorePatterns, filteredDirents, files, output;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, (0, promises_1.readdir)(dir, { withFileTypes: true })];
                case 1:
                    dirents = _a.sent();
                    updatedIgnorePatterns = Array.isArray(ignorePatterns) && ignorePatterns.length > 0 ? ignorePatterns : ["node_modules"];
                    filteredDirents = dirents.filter(function (_a) {
                        var name = _a.name;
                        return updatedIgnorePatterns.some(function (pattern) { return (0, minimatch_1.minimatch)(name, pattern) === true; }) === false;
                    });
                    return [4, Promise.all(filteredDirents.map(function (dirent) {
                            var res = (0, path_1.join)(dir, dirent.name);
                            return dirent.isDirectory() ? readFileNames(res) : res;
                        }))];
                case 2:
                    files = _a.sent();
                    output = files.flat();
                    return [2, output];
            }
        });
    });
}
exports.readFileNames = readFileNames;
function loadSourceFiles(_a) {
    var ignorePatterns = _a.ignorePatterns, project = _a.project, projectPath = _a.projectPath;
    return __awaiter(this, void 0, void 0, function () {
        var allowedFiles, files;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    allowedFiles = [".ts", ".js", ".tsx", ".jsx", ".mjs", ".cjs", ".mts"];
                    return [4, readFileNames(projectPath, ignorePatterns)];
                case 1:
                    files = (_b.sent()).filter(function (file) {
                        return allowedFiles.includes((0, path_1.extname)(file));
                    });
                    files.forEach(function (filePath) {
                        project.addSourceFileAtPathIfExists(filePath);
                    });
                    return [2];
            }
        });
    });
}
function createProject(projectPath, ignorePatterns) {
    return __awaiter(this, void 0, void 0, function () {
        var project;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    project = getProjectInstance();
                    return [4, loadSourceFiles({
                            ignorePatterns: ignorePatterns,
                            project: project,
                            projectPath: projectPath !== null && projectPath !== void 0 ? projectPath : (0, path_1.join)(process.cwd(), "src"),
                        })];
                case 1:
                    _a.sent();
                    return [2, project];
            }
        });
    });
}
exports.createProject = createProject;
function extractComponentMetadata(classDeclaration) {
    var componentMetadata = {
        designSystems: [],
        filewiseOccurences: {},
        id: '',
        isSelfDeclared: true,
        name: classDeclaration.getName(),
        tags: [],
        totalOccurences: 0,
    };
    classDeclaration.getDecorators().forEach(function (decorator) {
        var decoratorName = decorator.getName();
        if (decoratorName === 'Component') {
            var args = decorator.getArguments();
            if (args.length > 0) {
                args[0].getProperties().forEach(function (prop) {
                    var _a;
                    if (prop.getName() === 'tag') {
                        componentMetadata.selector = (_a = prop.getInitializer()) === null || _a === void 0 ? void 0 : _a.getText().replace(/'/g, '');
                    }
                });
            }
        }
    });
    return componentMetadata;
}
function getFileWiseLocations(_a) {
    var sourceFiles = _a.sourceFiles, selector = _a.selector, project = _a.project;
    var filewiseLocation = {};
    sourceFiles.forEach(function (sourceFile) {
        var filePath = (0, names_1.getPathRelativeToRoot)(sourceFile, project);
        function processJsxElement(jsxElement, isSelfClosing) {
            if (isSelfClosing === void 0) { isSelfClosing = false; }
            var name = (0, names_1.getTagName)(isSelfClosing ? jsxElement : jsxElement.getOpeningElement());
            var _a = sourceFile.getLineAndColumnAtPos(jsxElement.getStart()), line = _a.line, column = _a.column;
            if (name === selector) {
                if (!filewiseLocation[filePath]) {
                    filewiseLocation[filePath] = [];
                }
                filewiseLocation[filePath].push({
                    lineNumber: line,
                    columnNumber: column,
                });
            }
        }
        sourceFile.getDescendantsOfKind(ts_morph_1.SyntaxKind.JsxElement).forEach(function (jsxElement) {
            processJsxElement(jsxElement);
        });
        sourceFile.getDescendantsOfKind(ts_morph_1.SyntaxKind.JsxSelfClosingElement).forEach(function (jsxSelfClosingElement) {
            processJsxElement(jsxSelfClosingElement, true);
        });
    });
    return filewiseLocation;
}
function parseStencil(_a) {
    var project = _a.project, ignorePatterns = _a.ignorePatterns;
    return __awaiter(this, void 0, void 0, function () {
        var projectInstance, sourceFiles;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4, createProject(project, ignorePatterns)];
                case 1:
                    projectInstance = _b.sent();
                    sourceFiles = projectInstance.getSourceFiles();
                    sourceFiles.forEach(function (sourceFile) {
                        sourceFile.getClasses().forEach(function (classD) {
                            var meta = extractComponentMetadata(classD);
                            if (meta && meta.selector) {
                                var filePath = (0, names_1.getPathRelativeToRoot)(sourceFile, project);
                                meta.id = "".concat(filePath, "/").concat(meta.name);
                                meta.filewiseLocation = getFileWiseLocations({ sourceFiles: sourceFiles, project: project, selector: meta.selector });
                                meta.filewiseOccurences = {};
                                Object.keys(meta.filewiseLocation).forEach(function (loc) {
                                    meta.filewiseOccurences[loc] = meta.filewiseLocation[loc].length;
                                });
                                console.log(meta);
                            }
                        });
                    });
                    return [2, {}];
            }
        });
    });
}
exports.parseStencil = parseStencil;
