// @componlyco/berkoukes/dist/tsc/components/MultiDropdown/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { MultiDropdown } from "./index";
import { Button } from '../Button';
export default {
    title: "MultiDropdown",
    component: MultiDropdown,
    args: {
        items: [
            {
                title: '@componly',
                items: []
            },
            {
                title: '@mui',
                items: [
                    {
                        title: '@mui/material',
                        items: [
                            {
                                title: '@mui/material@5^11',
                                value: '@mui/material@5^11',
                            },
                            {
                                title: '@mui/material@5^12',
                                value: '@mui/material@5^12',
                            }
                        ]
                    },
                    {
                        title: '@mui/core',
                        items: [
                            {
                                title: '@mui/core@5^11',
                                value: '@mui/core@5^11',
                            },
                            {
                                title: '@mui/core@5^12',
                                value: '@mui/core@5^12',
                            }
                        ]
                    }
                ]
            },
            {
                title: '@berkoukes',
                items: [
                    {
                        title: '@berkoukes/core',
                        items: [
                            {
                                title: '@berkoukes/core@5^11',
                                value: '@berkoukes/core@5^11',
                            },
                            {
                                title: '@berkoukes/core@5^12',
                                value: '@berkoukes/core@5^12'
                            }
                        ]
                    }
                ]
            }
        ],
        onConfirm: function (val) { return console.log(val); },
        width: 500
    }
};
export var Template = function (props) {
    return _jsx(MultiDropdown, __assign({}, props, { children: _jsx(Button, { label: "Target", type: "primary" }) }));
};
