// @componlyco/core/dist/tsc/parse/angular/compiler/src/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.preserveWhitespacesDefault = exports.CompilerConfig = void 0;
var core_1 = require("./core");
var util_1 = require("./util");
var CompilerConfig = (function () {
    function CompilerConfig(_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.defaultEncapsulation, defaultEncapsulation = _c === void 0 ? core_1.ViewEncapsulation.Emulated : _c, _d = _b.useJit, useJit = _d === void 0 ? true : _d, _e = _b.missingTranslation, missingTranslation = _e === void 0 ? null : _e, preserveWhitespaces = _b.preserveWhitespaces, strictInjectionParameters = _b.strictInjectionParameters;
        this.defaultEncapsulation = defaultEncapsulation;
        this.useJit = !!useJit;
        this.missingTranslation = missingTranslation;
        this.preserveWhitespaces = preserveWhitespacesDefault((0, util_1.noUndefined)(preserveWhitespaces));
        this.strictInjectionParameters = strictInjectionParameters === true;
    }
    return CompilerConfig;
}());
exports.CompilerConfig = CompilerConfig;
function preserveWhitespacesDefault(preserveWhitespacesOption, defaultSetting) {
    if (defaultSetting === void 0) { defaultSetting = false; }
    return preserveWhitespacesOption === null ? defaultSetting : preserveWhitespacesOption;
}
exports.preserveWhitespacesDefault = preserveWhitespacesDefault;
