// @componlyco/berkoukes/dist/tsc/components/MetricItem/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from "react";
import { MetricItem } from "./index";
import { Button } from "../Button";
export default {
    title: "MetricItem",
    component: MetricItem,
    args: {
        loading: false,
        maxWidth: 600,
        title: 'Component Name',
        onClose: function () { return null; },
        tags: [
            'Core'
        ],
        library: { title: 'Material UI', logo: null },
        tabsValue: "Page",
        tabsItems: [
            { text: "Page" },
            { text: "Project" },
            { text: "All projects" }
        ],
        pieChartLegend: [
            {
                "name": "deprecated: 80%",
                "color": "#007BFF"
            },
            {
                "name": "legacy: 20%",
                "color": "#B692F6"
            },
            {
                "name": "core: 0%",
                "color": "#FF7262"
            }
        ],
        pieChart: {
            "tooltip": {
                "trigger": "item",
                "className": "chart-tooltip",
                "renderMode": "html",
                "appendToBody": true,
                "borderRadius": 30,
                "borderColor": "#F5F6F8",
                "borderWidth": 2,
                "boxShadow": "5px 10px #888888",
                "padding": 18,
                "axisPointer": {
                    "type": "cross"
                }
            },
            "legend": {
                "orient": "vertical",
                "left": "left",
                "show": false
            },
            "series": [
                {
                    "type": "pie",
                    "radius": [
                        "40%",
                        "70%"
                    ],
                    "avoidLabelOverlap": false,
                    "label": {
                        "show": false,
                        "position": "center"
                    },
                    "emphasis": {
                        "label": {
                            "show": true,
                            "fontSize": 16,
                            "color": "#000",
                            "stroke": "#fff",
                            "fontWeight": "bold",
                            "fontFamily": "Ruda"
                        }
                    },
                    "labelLine": {
                        "show": false
                    },
                    "data": [
                        {
                            "name": "deprecated",
                            "value": 149,
                            "itemStyle": {
                                "color": "#007BFF"
                            }
                        },
                        {
                            "name": "legacy",
                            "value": 27,
                            "itemStyle": {
                                "color": "#B692F6"
                            }
                        },
                        {
                            "name": "core",
                            "value": 0,
                            "itemStyle": {
                                "color": "#FF7262"
                            }
                        }
                    ]
                }
            ]
        },
        metricsCards: [
            {
                value: 23,
                primaryText: 'Usage',
                onClick: function () { return alert('OPEN FULL REPORT'); }
            },
            {
                primaryText: 'Present on how many files',
                value: 12,
                onClick: function () { return alert('OPEN FULL REPORT'); }
            }
        ],
        location: {
            path: 'src/assets/icons/CheckIcon.tsx',
            column: '19',
            line: '3'
        },
        brands: [
            {
                "alias": "SpaceY",
                "logo": "https://componly-staging.s3.eu-central-1.wasabisys.com/SPaCEY.svg",
            },
            {
                "alias": "Tezla",
                "logo": "https://componly-staging.s3.eu-central-1.wasabisys.com/TEZLA.svg",
            }
        ],
        projects: [
            {
                "name": "Webapp",
            }
        ],
        overrides: {
            data: [
                {
                    "name": "@mui/material/Typography",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Box",
                    "occurence": 1
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                }
            ],
            totalOccurrence: 4,
            page_size: 5,
            hasNextPage: true,
            setPage: function () { return null; }
        },
        overridedBy: {
            data: [
                {
                    "name": "@mui/material/Typography",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Box",
                    "occurence": 1
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                },
                {
                    "name": "@mui/material/Stack",
                    "occurence": 2
                }
            ],
            totalOccurrence: 4,
            page_size: 5,
            hasNextPage: true,
            setPage: function () { return null; }
        }
    }
};
export var Template = function (args) {
    var _a = useState(false), isOpened = _a[0], setOpen = _a[1];
    return _jsxs(_Fragment, { children: [_jsx(Button, { type: "primary", label: "Modal", onClick: function () { return setOpen(true); } }), _jsx(MetricItem, __assign({}, args, { onClose: function () { return setOpen(false); }, isOpened: isOpened }))] });
};
