// @componlyco/berkoukes/dist/tsc/components/Card/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Card } from "./index";
export default {
    title: "Card",
    component: Card,
};
var Template = function (args) { return _jsx(Card, __assign({}, args)); };
export var Simple = function (args) { return (_jsx("div", __assign({ style: {
        width: "50vw",
        height: "50vh",
    } }, { children: _jsx(Card, __assign({}, args)) }))); };
export var Default = Template.bind({});
Default.args = {
    primaryText: "Standardization of design tokens across a design system to maximize consistency.",
    secondaryText: "Standardization of design tokens across a design system to maximize consistency.",
    variant: "super-charge",
};
export var Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
    primaryText: "Standardization of design tokens across a design system to maximize consistency.",
    secondaryText: "Standardization of design tokens across a design system to maximize consistency.",
    variant: "super-charge",
};
