// @componlyco/schemas/dist/tsc/figma.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FigmaScansSchema = exports.FigmaScanSchema = exports.FigmaDesignSystemComponentsSchema = exports.FigmaDesignSystemSchema = exports.FigmaInstanceSchema = exports.FigmaVariantSchema = exports.FigmaPropSchema = exports.FigmaUsageSchema = exports.FigmaStyleSchema = void 0;
const zod_1 = require("zod");
exports.FigmaStyleSchema = zod_1.z.object({
    id: zod_1.z.string(),
    key: zod_1.z.string(),
    name: zod_1.z.string(),
    styleType: zod_1.z.string(),
    remote: zod_1.z.boolean(),
    description: zod_1.z.string().optional(),
    count: zod_1.z.number(),
    designSystem: zod_1.z.string().optional(),
});
exports.FigmaUsageSchema = zod_1.z.object({
    node: zod_1.z.string(),
    page: zod_1.z.string(),
    frame: zod_1.z.string(),
    link: zod_1.z.string(),
});
exports.FigmaPropSchema = zod_1.z.object({
    name: zod_1.z.string(),
    count: zod_1.z.number(),
    type: zod_1.z.string(),
});
exports.FigmaVariantSchema = zod_1.z.object({
    name: zod_1.z.string(),
    value: zod_1.z.union([zod_1.z.string(), zod_1.z.boolean()]),
    count: zod_1.z.number(),
    type: zod_1.z.string(),
});
exports.FigmaInstanceSchema = zod_1.z.object({
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    count: zod_1.z.number(),
    componentId: zod_1.z.string().nullable(),
    props: zod_1.z.record(exports.FigmaPropSchema).nullable(),
    variants: zod_1.z.record(exports.FigmaVariantSchema).nullable(),
    styles: zod_1.z.record(exports.FigmaStyleSchema).nullable(),
    designSystem: zod_1.z.string().nullable(),
    usage: zod_1.z.array(exports.FigmaUsageSchema),
    debt: zod_1.z.array(zod_1.z.any()),
    frame: zod_1.z.boolean().optional()
});
exports.FigmaDesignSystemSchema = zod_1.z.object({
    alias: zod_1.z.string(),
});
exports.FigmaDesignSystemComponentsSchema = zod_1.z.array(zod_1.z.object({
    name: zod_1.z.string(),
    components: zod_1.z.array(zod_1.z.object({
        name: zod_1.z.string(),
        props: zod_1.z.array(zod_1.z.object({
            name: zod_1.z.string(),
            values: zod_1.z.array(zod_1.z.string()).optional()
        })).optional()
    }))
}));
exports.FigmaScanSchema = zod_1.z.object({
    instances: zod_1.z.record(exports.FigmaInstanceSchema),
    styles: zod_1.z.record(exports.FigmaStyleSchema),
    designSystems: zod_1.z.array(exports.FigmaDesignSystemSchema),
    timestamp: zod_1.z.string(),
    fileName: zod_1.z.string(),
    fileKey: zod_1.z.string(),
    dsComponents: exports.FigmaDesignSystemComponentsSchema
});
exports.FigmaScansSchema = zod_1.z.array(exports.FigmaScanSchema);
