// @componlyco/berkoukes/dist/tsc/components/SecondaryHeader/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { ButtonType } from "../Button";
import { SecondaryHeader } from "./index";
export default {
    title: "SecondaryHeader",
    component: SecondaryHeader,
    args: {
        pageName: "Page name",
        tabs: [
            {
                type: ButtonType.Secondary,
                label: "Step1/6",
            },
            {
                type: ButtonType.NoBorders,
                label: "Step2/6",
            },
            {
                type: ButtonType.NoBorders,
                label: "Step3/6",
            },
            {
                type: ButtonType.NoBorders,
                label: "Step4/6",
            },
            {
                type: ButtonType.NoBorders,
                label: "Step5/6",
            },
            {
                type: ButtonType.NoBorders,
                label: "Step6/6",
            },
        ],
    },
};
export var Template = function (args) { return (_jsx(SecondaryHeader, __assign({}, args))); };
