// @componlyco/berkoukes/dist/tsc/components/MultiSelect/index.stories.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Avatar } from "../Avatar";
import { MultiSelect } from "./index";
export default {
    title: "MultiSelect",
    component: MultiSelect,
    args: {
        data: [
            {
                value: "react",
                label: "React",
                icons: [
                    _jsx(Avatar, { shape: "circle", type: "image", size: "xs", src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" }),
                ],
            },
            {
                value: "ng",
                label: "Angular",
                icons: [
                    _jsx(Avatar, { shape: "circle", type: "image", size: "xs", src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" }),
                ],
            },
            {
                value: "svelte",
                label: "Svelte",
                icons: [
                    _jsx(Avatar, { shape: "circle", type: "image", size: "xs", src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" }),
                ],
            },
            {
                value: "vue",
                label: "Vue",
                icons: [
                    _jsx(Avatar, { shape: "circle", type: "image", size: "xs", src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" }),
                ],
            },
            {
                value: "next",
                label: "Next.js",
                icons: [
                    _jsx(Avatar, { shape: "circle", type: "image", size: "xs", src: "https://images.unsplash.com/photo-1535713875002-d1d0cf377fde?ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=250&q=80" }),
                ],
            },
        ],
        label: "Label",
        placeholder: "Select Multiple item",
        optionalText: "Optional",
        helperText: "Helper text",
        helpTooltipProps: {
            position: "top",
            title: "Header",
            description: "In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content.",
        }
    },
};
var Template = function (args) { return _jsx(MultiSelect, __assign({}, args)); };
export var Default = Template.bind({});
export var Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
};
export var Error = Template.bind({});
Error.args = {
    error: "Pick at least one item.",
};
export var WithoutOptional = Template.bind({});
WithoutOptional.args = {
    optionalText: undefined,
};
export var Searchable = Template.bind({});
Searchable.args = {
    searchable: true,
};
